<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>preciseTADworkshop • preciseTADworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="preciseTADworkshop">
<meta property="og:description" content="Chromosome conformation capture technologies combined with
    high-throughput sequencing (Hi-C) have revealed that chromatin undergoes 
    layers of compaction through DNA looping and folding, forming dynamic 3D 
    structures. Among these are Topologically Associating Domains (TADs), which 
    are known to play critical roles in cell dynamics like gene regulation and 
    cell differentiation. Precise identification of TAD boundaries (TAD-calling)
    remains difficult, as it is strongly reliant on Hi-C data resolution. 
    Obtaining genome-wide chromatin interactions at high-resolution is costly 
    resulting in low resolution of Hi-C matrices and high uncertainty in the 
    location of domain boundaries. In this workshop, we will introduce a novel 
    approach for transforming TAD-calling into a supervised machine learning 
    framework using functional genomic elements. We will walk the user through
    the customizable model building process and finish with an application of
    precisely predicting boundaries at base-level resolution. The methods 
    demonstrated in this workshop are from the `preciseTAD` R package.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">preciseTADworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/preciseTAD_tutorial.html">preciseTAD: A machine learning framework for precise 3D domain boundary prediction at base-level resolution</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dozmorovlab/preciseTADworkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<div id="precisetad-a-machine-learning-framework-for-precise-3d-domain-boundary-prediction-at-base-level-resolution" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#precisetad-a-machine-learning-framework-for-precise-3d-domain-boundary-prediction-at-base-level-resolution" class="anchor"></a>preciseTAD: A machine learning framework for precise 3D domain boundary prediction at base-level resolution</h1></div>
<p>The <a href="https://github.com/dozmorovlab/preciseTADworkshop">preciseTADworkshop</a> introduces methods for transforming the identification of boundaries that demarcate Topologically Associating Domains (TADs)– referred to as TAD-calling–into a supervised machine learning framework. Chromosome conformation capture technologies combined with high-throughput sequencing (Hi-C) have revealed that chromatin undergoes layers of compaction through DNA looping and folding, forming dynamic 3-dimensional (3D) structures. Among these are TADs, which are known to play critical roles in cell dynamics like gene regulation and cell differentiation. However, precise TAD-calling remains difficult, as it is strongly reliant on Hi-C data resolution. Obtaining genome-wide chromatin interactions at high-resolution is costly resulting in low resolution of Hi-C matrices and high uncertainty in the location of domain boundaries. In this workshop we will circumvent this resolution restriction by building predictive models that leverage high resolution functional genomic element data (ChIP-seq). As an application, we will demonstrate that these methods provide more precise boundary detection compared to a conventional TAD-calling algorithm by evaluating a variety of visualization techniques in relation to the enrichment of key molecular drivers of 3D chromatin. The methods discussed in this workshop will give users tools for bridging the resolution gap between 1D ChIP-seq annotations and 3D Hi-C sequencing data for more precise and biologically meaningful boundary identification.</p>
<p>This workshop is based on <a href="https://doi.org/10.1101/2020.09.03.282186">Spiro C. Stilianoudakis, Mikhail G. Dozmorov; “preciseTAD: A machine learning framework for precise 3D domain boundary prediction at base-level resolution”</a>.</p>
<p>Key materials for the workshop:</p>
<ul>
<li><a href="https://stilianoudakis.github.io/slides_preciseTADworkshop/#1">Slides with a brief introduction for the workshop</a></li>
<li><a href="https://github.com/dozmorovlab/preciseTADworkshop">preciseTADworkshop GitHub repo</a></li>
<li><a href="https://hub.docker.com/repository/docker/stilianoudakis/precisetadworkshop">preciseTADworkshop Docker image</a></li>
<li><a href="https://dozmorovlab.github.io/preciseTADworkshop/">preciseTADworkshop pkgdown website</a></li>
</ul>
<p>This workshop will be presented at the <a href="https://eurobioc2020.bioconductor.org/">European Bioconductor Virtual Conference 2020</a>, December 6th, 2020, 5:30pm-6:15pm</p>
</div>
<div id="using-docker" class="section level1">
<h1 class="hasAnchor">
<a href="#using-docker" class="anchor"></a>Using Docker</h1>
<ul>
<li>Pull the latest version of the <code>preciseTADworkshop</code> Docker image, <code>docker pull stilianoudakis/precisetadworkshop:latest</code>
</li>
<li>Run <code>docker run -e PASSWORD=yourpassword -p 8787:8787 -d --rm stilianoudakis/precisetadworkshop:latest</code>. Use <code>-v $(pwd):/home/rstudio</code> argument to map your local directory to the container.</li>
<li>Log in to RStudio at <a href="http://localhost:8787">http://localhost:8787</a> using username <code>rstudio</code> and password <code>yourpassword</code>. Note that on Windows you need to provide your localhost IP address like <code>http://191.163.92.108:8787/</code> - find it using <code>docker-machine ip default</code> in Docker’s terminal.</li>
<li>Run <code><a href="https://rdrr.io/r/utils/browseVignettes.html">browseVignettes(package = "preciseTADworkshop")</a></code>. Click on one of the links, “HTML”, “source”, “R code”.</li>
</ul>
</div>
<div id="local-installation" class="section level1">
<h1 class="hasAnchor">
<a href="#local-installation" class="anchor"></a>Local Installation</h1>
<pre><code>if(!require(devtools)) install.packages("devtools")
devtools::install_github(repo = "dozmorovlab/preciseTADworkshop", build_vignettes = TRUE)</code></pre>
<p>If installation fails due to missing packages, install them as follows:</p>
<pre><code>if(!require(BiocManager)) install.packages("BiocManager")
BiocManager::install(c('preciseTAD', 'preciseTADhub'))</code></pre>
</div>
<div id="workshop-description" class="section level1">
<h1 class="hasAnchor">
<a href="#workshop-description" class="anchor"></a>Workshop Description</h1>
<div id="contact-information" class="section level2">
<h2 class="hasAnchor">
<a href="#contact-information" class="anchor"></a>Contact information</h2>
<ul>
<li>Spiro Stilianoudakis (<a href="mailto:stilianoudasc@vcu.edu" class="email">stilianoudasc@vcu.edu</a>)</li>
<li>Mikhail Dozmorov (<a href="mailto:mikhail.dozmorov@vcuhealth.org" class="email">mikhail.dozmorov@vcuhealth.org</a>)</li>
</ul>
</div>
<div id="pre-requisites" class="section level2">
<h2 class="hasAnchor">
<a href="#pre-requisites" class="anchor"></a>Pre-requisites</h2>
<ul>
<li>Basic knowledge of R syntax and command-line tools</li>
<li>Familiarity with Hi-C chromatin conformation capture technology</li>
<li>Understanding of Hi-C data properties (contact matrices, interaction frequencies, Topologically Associating Domains, etc)</li>
<li>Familiarity with TAD-callers, specifically <a href="https://github.com/aidenlab/juicer/wiki/Arrowhead">Arrowhead</a>
</li>
<li>Familiarity with supervised machine learning techniques (e.g. classification) and clustering algorithms</li>
</ul>
</div>
</div>
<div id="r--bioconductor-packages-used" class="section level1">
<h1 class="hasAnchor">
<a href="#r--bioconductor-packages-used" class="anchor"></a>R / Bioconductor packages used</h1>
<ul>
<li><a href="https://www.bioconductor.org/packages/preciseTAD">preciseTAD</a></li>
<li><a href="https://github.com/dozmorovlab/preciseTADhub">preciseTADhub</a></li>
</ul>
<div id="time-outline" class="section level2">
<h2 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h2>
<p>The workshop duration is ~45 min. The approximate timing of activities are shown below:</p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Overview of <em>preciseTAD</em>
</td>
<td>10m</td>
</tr>
<tr class="even">
<td>Model building</td>
<td>15m</td>
</tr>
<tr class="odd">
<td>Precise domain boundary prediction</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Questions and answers session</td>
<td>5m</td>
</tr>
</tbody>
</table>
<div id="learning-goals" class="section level3">
<h3 class="hasAnchor">
<a href="#learning-goals" class="anchor"></a>Learning Goals</h3>
<ul>
<li>Get familiar with Hi-C data</li>
<li>Understand to transform TAD boundary prediction into a supervised machine learning framework</li>
<li>Build a predictive model using functional genomic elements</li>
<li>Predict TAD boundaries at base-level resolution</li>
<li>Understand and compare results using a variety of different vizualization techniques including enriched heatmaps, signal profile plots, and venn diagrams</li>
</ul>
</div>
</div>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/mdozmorov/preciseTADworkshop/">https://​github.com/​mdozmorov/​preciseTADworkshop/​</a>
</li>
<li>Report a bug at <br><a href="https://github.com/mdozmorov/preciseTADworkshop/issues/new/choose">https://​github.com/​mdozmorov/​preciseTADworkshop/​issues/​new/​choose</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Spiro Stilianoudakis <br><small class="roles"> Author, maintainer </small>  </li>
<li>Mikhail Dozmorov <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-0086-8358" target="orcid.widget" aria-label="ORCID"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p>Developed by Spiro Stilianoudakis, Mikhail Dozmorov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
