# preciseTADworkshop

Precise prediction of TAD/loop boundaries, https://doi.org/10.1101/2020.09.03.282186


# European Bioconductor Conference 2020
https://openreview.net/group?id=bioconductor.org/EuroBioC/2020/Conference
Abstract submission for talks, posters, and workshops.

Virtual  Dec 14 2020  https://eurobioc2020.bioconductor.org/  eurobioc2020@stat.unipd.it

Please see the venue website for more information.

Submission Start: Aug 31 2020 12:00AM UTC-0, End: Oct 30 2020 12:00AM UTC-0

Add: 

**Title**

A workshop for <!--leveraging functional genomic annotations to--> precisely predicting 3D domain boundaries at base-level resolution using a custom built machine learning framework

**Authors**

Search author profile by first, middle and last name or email address. If the profile is not found, you can add the author completing first, middle, last and name and author email address.

Spiro Stilianoudakis stilianoudasc@mymail.vcu.edu  
Mikhail G. Dozmorov mikhail.dozmorov@vcuhealth.org

**Keywords**

Hi-C, TAD, machine learning, classification, random forest, clustering, DBSCAN

**TL;DR**

In this workshop we outline novel methods for transforming TAD-calling into a supervised machine learning framework by leveraging high resolution functional genomic annotations to precisely predict TAD-boundaries at base-level resolution.

**Abstract**

High-throughput chromosome conformation capture technology (Hi-C) has revealed extensive DNA looping and folding into discrete 3D structures referred to as Topologically Associating Domains (TADs).  TADs are critical for cellular processes like gene regulation and cell differentiation. The relatively low resolution of Hi-C data (regions of several kilobases in size) prevents precise mapping of TAD boundaries by conventional TAD-callers. In contrast, the enrichment of high resolution functional genomic annotations (ChIP-seq) at boundaries offers a computational approach toward improved precision of boundary identification. The workshop presented here describes how to transform TAD-calling into a supervised machine learning framework. We describe how to build and optimize random forest classifers that leverage known molecular drivers of 3D chromatin in order to predict TAD-boundary regions. We then describe an algorithm that utilizes density-based clustering (DBSCAN) and partitioning around medoids (PAM) to precisely predict boundaries at base-level resolution using built-in random forest models. Finally, we compare our results with the popular Arrowhead TAD-caller using both visual and statistical methods. The bulk of this workshop uses the R programming environment and is an extension of the preciseTAD R package. The processes described can be performed with most operating systems and a single computer.

**Affiliation**

List your affiliation

**Gender**

**Bioconductor Package Maintenance**

preciseTAD

**Associated Packages**

preciseTAD

**Submission Type**

- Short workshop (30 - 45 minutes) such as a demonstration of a package or a workflow

**Submission Is Not Registration**

I am aware that this form does not constitute a registration. See the conference page for details about the registration.

**Code Of Conduct**

I agree to abide by the Bioconductor Code of Conduct (https://eurobioc2020.bioconductor.org/code_of_conduct) during the conference and in all interactions with the Bioconductor community. This includes prohibitions on harassment, discrimination, and intellectual bullying. It also includes a social media policy that anything presented at the conference may be posted publicly unless explicitly requested otherwise by the presenter.
