---
title: "preciseTAD: a machine-learning framework for predicting boundaries of 3D genomic elements"
author:
- name: Spiro Stilianoudakis
  affiliation:
  - &1 Department of Biostatistics, Virginia Commonwealth University, Richmond, VA
- name: Mikhail Dozmorov
  affiliation:
  - *1
date: '`r format(Sys.Date(), "%B %e, %Y")`'
output:
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    toc_depth: 3
    fig_width: 5
bibliography: references.bib
vignette: >
  %\VignetteIndexEntry{preciseTAD: a machine-learning framework for predicting boundaries of 3D genomic elements}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE}
# Set up the environment
knitr::opts_chunk$set(echo=T, warnings=F, message=F,
  collapse = TRUE,
  comment = "#>"
)
set.seed(1)
options(stringsAsFactors = FALSE)
```

## Installation

```{r}
# if (!requireNamespace("BiocManager", quietly=TRUE))
#     install.packages("BiocManager")
# BiocManager::install("preciseTAD")
library(preciseTAD)
```

## Introduction

The advent of chromosome conformation capture (3C) sequencing technologies, and its successor Hi-C, have revealed a hierarchy of the 3-dimensional (3D) structure of the human genome such as chromatin loops [@rao20143d], Topologically Associating Domains (TADs) [@nora2012spatial;@dixon2012topological], and A/B compartments [@lieberman2009comprehensive]. At the megabase scale, TADs represent regions on the linear genome that are highly self-interacting. Emerging evidence has linked TADs to critical roles in cell dynamics and cell differentiation. Studies have shown that TADs themselves are highly conserved across species and cell lines [@dixon2012topological; @krefting2018evolutionary; @nora2012spatial; @dixon2015chromatin; @rao20143d; @Schmitt2016aa]. Disruption of boundaries demarcating TADs and loops promotes cancer [@hnisz2016activation; @taberlay2016three] and other disorders [@lupianez2016breaking;@Sun:2018aa;@Meaburn:2007aa]. Identifying the precise location of TAD boundaries remains a top priority in our goal to fully understand the functionality of the human genome.

In this workshop, we present *preciseTAD*, an optimally tuned machine learning framework for precise identification of domain boundaries using genome annotation data. Our method utilizes the random forest (RF) algorithm trained on high-resolution chromatin state (BroadHMM), histone modification (HM), and transcription factor binding site (TFBS) data to predict low-resolution boundaries. We introduce a systematic pipeline for building the optimal boundary region prediction classifier. Translated from Hi-C data resolution level to base level (annotating each base and predicting its boundary probability), *preciseTAD* employs density-based clustering (DBSCAN) and partitioning around medoids (PAM) to detect genome annotation-guided boundary regions and points at a base-level resolution. This approach circumvents resolution restrictions of Hi-C data, allowing for the precise detection of biologically meaningful boundaries.

In this workshop we will demonstrate the following:

<!--
- Spatial associations (linear distance) between boundaries and annotations perform best for predicting boundary regions
- Simple random under-sampling effectively addresses the negative effect of class imbalance
-->
- *preciseTAD* predictions are more enriched for known molecular drivers of 3D chromatin, including CTCF, RAD21, SMC3, and ZNF143
- *preciseTAD*-predicted boundaries are more conserved across cell lines
- Using methods developed, a model trained in one cell line can predict boundaries in another cell line

## Obtaining ground truth TAD boundaries

*preciseTAD* uses pre-defined TAD boundary coordinates as ground truth. We opt to use the Arrowhead to define our ground truth TAD boundaries, a gold-standard TAD-caller developed by [Aiden Lab](https://github.com/aidenlab/juicer/wiki/Citing-Juicer) [@durand2016juicer]. The .hic files are publically available on GEO [here](), and are provided by the same team that developed Arrowhead [@rao20143d]. For the *preciseTAD* paper and this workshop we will be working with the "GSE63525_GM12878_insitu_primary.hic" file. This .hic file is a highly compressed binary file that has contact matrices from multiple biological replicate HiC experiments, at different resolutions, for GM12878. You can read the full protocol used to derive the various .hic files [here](https://www.cell.com/cell/pdfExtended/S0092-8674(14)01497-4).

Once you have the .hic file downloaded, you are ready to implement Arrowhead to call TAD boundaries. Below is an example of how you would perform this for every chromosome:

```{bash, eval=FALSE}
for i in 5000 10000 25000 50000 100000;
	do 
		echo $i.kb
			for j in {1..22};
				do
				echo chr$j
				java -jar juicebox_tools_7.5.jar arrowhead -c $j \ #chromosome to call TADs on
				                                           -r $i \ #HiC data resolution
				                                           ~/GSE63525_GM12878_insitu_primary.hic \ #location of the .HIC file
				                                           ~/arrowhead_output/GM12878/$i.b/chr$j #location to store the output
			done;
	done;
	
```

Note: This creates separate files for each chromosome (given a specific resolution). As an example, here is what the first few lines of CHR1 look like:

| chr1 | x1        | x2        | chr2 | y1        | y2        | color     | score               | uVarScore           | lVarScore           | upSign             | loSign              |
|------|-----------|-----------|------|-----------|-----------|-----------|---------------------|---------------------|---------------------|--------------------|---------------------|
| 1    | 49375000  | 50805000  | 1    | 49375000  | 50805000  | 255,255,0 | 0.8635365988485998  | 0.3025175910115416  | 0.2942087461534729  | 0.4431332556332556 | 0.6401515151515151  |
| 1    | 16830000  | 17230000  | 1    | 16830000  | 17230000  | 255,255,0 | 0.20513320424507786 | 0.3676005813255187  | 0.22358121923097696 | 0.4121951219512195 | 0.4792682926829268  |
| 1    | 163355000 | 164860000 | 1    | 163355000 | 164860000 | 255,255,0 | 0.8052794339967723  | 0.27908025641958345 | 0.3017420763866275  | 0.657032586290075  | 0.5470812683654226  |
| 1    | 231935000 | 233400000 | 1    | 231935000 | 233400000 | 255,255,0 | 0.43336944864061144 | 0.28052310457518204 | 0.27077299003747285 | 0.451432273589708  | 0.44314868804664725 |
| 1    | 149035000 | 149430000 | 1    | 149035000 | 149430000 | 255,255,0 | 0.8579084086781587  | 0.3477831683616397  | 0.18791010793617813 | 0.45375            | 0.805625            |

Explanations of each field are as follows:

- chromosome = the chromosome that the domain is located on
- x1,x2/y1,y2 = the interval spanned by the domain (contact domains manifest as squares on the diagonal of a Hi-C matrix and as such: x1=y1, x2=y2)
- color = the color that the feature will be rendered as if loaded in Juicebox
- corner_score = the corner score, a score indicating the likelihood that a pixel is at the corner of a contact domain. Higher values indicate a greater likelihood of being at the corner of a domain
- Uvar = the variance of the upper triangle
- Lvar = the variance of the lower triangle
- Usign = -1*(sum of the sign of the entries in the upper triangle)
- Lsign = sum of the sign of the entries in the lower triangle

For the purposes of this workshop, we are only interested in the first 3 columns, the chromosome, and the start and end coordinates of each called TAD. Since this is performed per chromosome, each chromosome specific file needs to be concatenated to obtain the full list of TADs on the entire genome. For convenience, the TADs called at 5kb resolution for GM12878 using Arrowhead are provided internally in the *preciseTAD* package.

```{r}
data("arrowhead_gm12878_5kb")
head(arrowhead_gm12878_5kb)
```

## Obtaining functional genomic elements

The next piece of data that we need are the functional genomic elements used to construction the feature space. These are processed narrowPeak ChIP-seq data in the form of .BED files with chromosome, and start and end coordinates of called peaks of signal enrichment. For this workshop we consider 3 classes of genomic elements: BroadHMM (N=15), histone modifications (N=11), and transcription factor binding sites (N=26). They are shown below:

| Genomic Class                               | Element         |
|---------------------------------------------|-----------------|
| BroadHMM chromatin segmentations            | 10TxnElongation |
|                                             | 11WeakTxn       |
|                                             | 12Repressed     |
|                                             | 13Heterochromlo |
|                                             | 14RepetitiveCNV |
|                                             | 15RepetitiveCNV |
|                                             | 1ActivePromoter |
|                                             | 2WeakPromoter   |
|                                             | 3PoisedPromoter |
|                                             | 4StrongEnhancer |
|                                             | 5StrongEnhancer |
|                                             | 6WeakEnhancer   |
|                                             | 7WeakEnhancer   |
|                                             | 8Insulator      |
|                                             | 9TxnTransition  |
| Histone Modifications (HM)                  | H2az            |
|                                             | H3k27ac         |
|                                             | H3k27me3        |
|                                             | H3k36me3        |
|                                             | H3k4me1         |
|                                             | H3k4me2         |
|                                             | H3k4me3         |
|                                             | H3k79me2        |
|                                             | H3k9ac          |
|                                             | H3k9me3         |
|                                             | H4k20me1        |
| Transcription Factor Binding Sites   (TFBS) | Atf3            |
|                                             | Cfos            |
|                                             | Cmyc            |
|                                             | Ctcf            |
|                                             | Egr1            |
|                                             | Ets1            |
|                                             | Gabp            |
|                                             | Jund            |
|                                             | Max             |
|                                             | Mazab85725      |
|                                             | Mef2a           |
|                                             | Mxi1            |
|                                             | P300            |
|                                             | Pmlsc71910      |
|                                             | Pol2            |
|                                             | Pu1             |
|                                             | Rad21           |
|                                             | Rfx5            |
|                                             | Sin3a           |
|                                             | Six5            |
|                                             | Smc3ab9263      |
|                                             | Sp1             |
|                                             | Srf             |
|                                             | Stat1           |
|                                             | Taf1            |
|                                             | Tr4             |
|                                             | Yy1             |
|                                             | Znf143          |

Cell type-specific genomic annotation data can be downloaded from the [ENCODE](http://www.encodeproject.org/chip-seq-matrix/?type=Experiment&replicates.library.biosample.donor.organism.scientific_name=Homo%20sapiens&assay_title=TF%20ChIP-seq&status=released) data portal as BED files. For convenience, we have included the list of elements in the `\data\genomicElements` folder of the github repository [here](https://github.com/dozmorovlab/preciseTADworkshop).

## Describing *preciseTAD* functionality

The training/testing data used for modeling is represented as a matrix with rows being genomic regions, columns being genomic annotations, and cells containing measures of association between them. Users have the option to concatenate genomic regions from multiple chromosomes.

To create the row-wise dimension of the data, `preciseTAD` uses *shifted binning*, a strategy for making the dimensions of the data matrix used for modeling by segmenting the linear genome into nonoverlapping regions. This step is transparent for the user. To create shifted bins, chromosome-specific bins start at half of the resolution *r*, and continue in congruent intervals of length *r* until the end of the chromosome (*mod r* + *r/2*), using hg19 genomic coordinates. The shifted genomic bins, are then defined as boundary regions (*Y = 1*) if they contain a called boundary, and non-boundary regions (*Y = 0*) otherwise, thus establishing the binary response vector (**Y**) used for classification (???Figure X). Intuitively, shifted bins are centered on borders between the original bins, thus capturing potential boundaries.

The column-wise dimension is formed by genomic annotations, such as transcription factor binding sites (TFBSs), histone modification marks, chromatin states. The *($log_{2}$) distances*, which enumerate the genomic distance from the center of each genomic bin to the center of the nearest ChIP-seq peak region of interest, form the feature space (???Figure X). The customized training/testing data formation offered by `preciseTAD` allows users to implement any binary classification machine learning algorithm. 

`preciseTAD` implements a random forest (RF) model, allowing tuning hyperparameters and applying feature reduction. The primary inputs are the training and testing data, a list of hyperparameter values, the number of cross-validation folds to use (if a grid of hyperparameter values is provided), and the metric used for optimization. The output includes the model object (necessary for downstream prediction of boundaries), the variable importance values, and a list of performance metrics when validating the model on the testing data (see Table 1). This model is then used to predict base-level precise boundary locations.

To predict the base-level location of domain boundaries, the `preciseTAD` model is applied for each base annotated with the aforementioned genomic annotations. The base-level probabilities are clustered using density-based clustering and scalable data partitioning techniques, to narrow boundary regions and points. First, the probability vector, $p_{n_{i}}$, is extracted ($n_{i}$ representing the length of chromosome $i$). Next, *DBSCAN* (Density-based Spatial Clustering of Applications with Noise) [@ester1996density; @hahsler2019dbscan] is applied to the matrix of pairwise genomic distances between bases with $p_{n_{i}} \ge t$, where $t$ is a threshold determined by the user. The resulting clusters of highly predictive bases identified by DBSCAN are termed *preciseTAD boundary regions* (PTBR). To precisely identify a single base among each PTBR, *preciseTAD* implements partitioning around medoids (PAM) within each cluster. The corresponding cluster medoid is defined as a *preciseTAD boundary point* (PTBP), making it the most representative and biologically meaningful base within each clustered PTBR. The output includes a list of genomic coordinates of PTBPs and PTBSs.

`preciseTAD` allows us to use the pre-trained model to predict the precise location of domain boundaries in cell types without Hi-C data but with genome annotation data. Specifically, only cell-type-specific ChIP-seq data (BED format) for CTCF, RAD21, SMC3, and ZNF143 transcription factor binding sites is required. We provide models pre-trained using GM12878 and K562 genome annotation data, and Arrowhead- and Peakachu-predicted boundaries for chromosome-specific prediction of precise domain boundaries in other cell types.

### Model building

### Optimization

### Boundary prediction with *preciseTAD*

### Using *preciseTAD* with Juicebox

### Cross-cell-type prediction

## References