<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>preciseTAD: A machine learning framework for precise 3D domain boundary prediction at base-level resolution • preciseTADworkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="preciseTAD: A machine learning framework for precise 3D domain boundary prediction at base-level resolution">
<meta property="og:description" content="preciseTADworkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">preciseTADworkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/preciseTAD_tutorial.html">preciseTAD: A machine learning framework for precise 3D domain boundary prediction at base-level resolution</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dozmorovlab/preciseTADworkshop">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="preciseTAD_tutorial_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="preciseTAD_tutorial_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="preciseTAD_tutorial_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>preciseTAD: A machine learning framework for precise 3D domain boundary prediction at base-level resolution</h1>
                        <h4 class="author">Spiro Stilianoudakis</h4>
            <address class="author_afil">
      Department of Biostatistics, Virginia Commonwealth University, Richmond, VA<br><h4 class="author">Mikhail Dozmorov</h4>
            <address class="author_afil">
      Department of Biostatistics, Virginia Commonwealth University, Richmond, VA<br><h4 class="date">December 8, 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/mdozmorov/preciseTADworkshop/blob/master/vignettes/preciseTAD_tutorial.Rmd"><code>vignettes/preciseTAD_tutorial.Rmd</code></a></small>
      <div class="hidden name"><code>preciseTAD_tutorial.Rmd</code></div>

    </address>
</address>
</div>

    
    
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://yihui.org/knitr/">knitr</a></span><span class="op">)</span>

<span class="co"># if (!requireNamespace("BiocManager", quietly=TRUE))</span>
<span class="co">#     install.packages("BiocManager")</span>
<span class="co"># BiocManager::install("preciseTAD")</span>
<span class="co">#library(preciseTAD)</span>

<span class="co">#library(devtools)</span>
<span class="co">#devtools::install_github("dozmorovlab/preciseTAD")</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/dozmorovlab/preciseTAD">preciseTAD</a></span><span class="op">)</span>

<span class="co">#install_github("stilianoudakis/preciseTADhub")</span>
<span class="co">#library(preciseTADhub)</span>
<span class="co">#library(ExperimentHub)</span></code></pre></div>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The advent of chromosome conformation capture (3C) sequencing technologies, and its successor Hi-C, have revealed a hierarchy of the 3-dimensional (3D) structure of the human genome such as chromatin loops <span class="citation">(Rao et al. 2014)</span>, Topologically Associating Domains (TADs) <span class="citation">(Nora et al. 2012; Dixon et al. 2012)</span>, and A/B compartments <span class="citation">(Lieberman-Aiden et al. 2009)</span>. At the megabase scale, TADs represent regions on the linear genome that are highly self-interacting. Emerging evidence has linked TADs to critical roles in cell dynamics and cell differentiation. Studies have shown that TADs themselves are highly conserved across species and cell lines <span class="citation">(Dixon et al. 2012, 2015; Krefting, Andrade-Navarro, and Ibn-Salem 2018; Nora et al. 2012; Rao et al. 2014; Schmitt et al. 2016)</span>. Disruption of boundaries demarcating TADs and loops promotes cancer <span class="citation">(Hnisz et al. 2016; Taberlay et al. 2016)</span> and other disorders <span class="citation">(Lupiáñez, Spielmann, and Mundlos 2016; Sun et al. 2018; Meaburn et al. 2007)</span>. Therefore, identifying the precise location of TAD boundaries remains a top priority in our goal to fully understand the functionality of the human genome.</p>
<p>In this workshop, we present <em>preciseTAD</em>, an optimally tuned machine learning framework for precise identification of domain boundaries using genome annotation data. Our method utilizes the random forest (RF) algorithm trained on high-resolution transcription factor binding site (TFBS) data to predict low-resolution boundaries. We introduce a systematic pipeline for building the optimal boundary region prediction classifier. Translated from Hi-C data resolution level to base level (annotating each base and predicting its boundary probability), <em>preciseTAD</em> employs density-based clustering (DBSCAN) and partitioning around medoids (PAM) to detect genome annotation-guided boundary regions and points at a base-level resolution. This approach circumvents resolution restrictions of Hi-C data, allowing for the precise detection of biologically meaningful boundaries.</p>
<p>In this workshop we will demonstrate the following:</p>
<!--
- Spatial associations (linear distance) between boundaries and annotations perform best for predicting boundary regions
- Simple random under-sampling effectively addresses the negative effect of class imbalance
-->
<ul>
<li>
<em>preciseTAD</em> predictions are more enriched for known molecular drivers of 3D chromatin, including CTCF, RAD21, SMC3, and ZNF143</li>
<li>
<em>preciseTAD</em>-predicted boundaries are more conserved across cell lines</li>
<li>Using the methods developed, a model trained in one cell line can accurately predict boundaries in another cell line</li>
</ul>
</div>
<div id="obtaining-ground-truth-tad-boundaries" class="section level1">
<h1 class="hasAnchor">
<a href="#obtaining-ground-truth-tad-boundaries" class="anchor"></a>Obtaining ground truth TAD boundaries</h1>
<p><em>preciseTAD</em> uses pre-defined TAD boundary coordinates as ground truth. We opt to use the Arrowhead algorithm to define our ground truth TAD boundaries, a gold-standard TAD-caller developed by <a href="https://github.com/aidenlab/juicer/wiki/Citing-Juicer">Aiden Lab</a> <span class="citation">(Durand, Shamim, et al. 2016)</span>. The .hic files are publically available on GEO <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE63525">here</a>, and are provided by the same team that developed Arrowhead <span class="citation">(Rao et al. 2014)</span>. For this workshop we will be working with the “GSE63525_GM12878_insitu_primary.hic” file. This .hic file is a highly compressed binary file that has contact matrices from multiple biological replicate HiC experiments, at different resolutions, for GM12878. You can read the full protocol used to derive the various .hic files <a href="https://www.cell.com/cell/pdfExtended/S0092-8674(14)01497-4">here</a>.</p>
<p>Once you have the .hic file downloaded, you are ready to implement Arrowhead to call TAD boundaries. Below is an example of how you would perform this for every chromosome and for a variety of resolutions:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">for</span> <span class="ex">i</span> in 5000 10000 25000 50000 100000<span class="kw">;</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>    <span class="kw">do</span> </span>
<span id="cb2-3"><a href="#cb2-3"></a>        <span class="bu">echo</span> <span class="va">$i</span>.kb</span>
<span id="cb2-4"><a href="#cb2-4"></a>            <span class="kw">for</span> <span class="ex">j</span> in <span class="dt">{1..22}</span><span class="kw">;</span></span>
<span id="cb2-5"><a href="#cb2-5"></a>                <span class="kw">do</span></span>
<span id="cb2-6"><a href="#cb2-6"></a>                <span class="bu">echo</span> chr<span class="va">$j</span></span>
<span id="cb2-7"><a href="#cb2-7"></a>                <span class="ex">java</span> -jar juicebox_tools_7.5.jar arrowhead -c <span class="va">$j</span> <span class="dt">\ </span>#chromosome to call TADs on</span>
<span id="cb2-8"><a href="#cb2-8"></a>                                                           <span class="ex">-r</span> <span class="va">$i</span> <span class="dt">\ </span>#HiC data resolution</span>
<span id="cb2-9"><a href="#cb2-9"></a>                                                           <span class="ex">~/GSE63525_GM12878_insitu_primary.hic</span> <span class="dt">\ </span>#location of the .HIC file</span>
<span id="cb2-10"><a href="#cb2-10"></a>                                                           <span class="ex">~/arrowhead_output/GM12878/</span><span class="va">$i</span><span class="ex">.b/chr</span><span class="va">$j</span> <span class="co">#location to store the output</span></span>
<span id="cb2-11"><a href="#cb2-11"></a>            <span class="kw">done</span>;</span>
<span id="cb2-12"><a href="#cb2-12"></a>    <span class="kw">done</span>;</span>
<span id="cb2-13"><a href="#cb2-13"></a>    </span></code></pre></div>
<p>Note: This creates separate files for each chromosome (given a specific resolution). As an example, here is what the first few lines of CHR1 look like, performed at 5kb:</p>
<table class="table">
<colgroup>
<col width="3%">
<col width="6%">
<col width="6%">
<col width="3%">
<col width="6%">
<col width="6%">
<col width="6%">
<col width="12%">
<col width="12%">
<col width="12%">
<col width="11%">
<col width="12%">
</colgroup>
<thead><tr class="header">
<th>chr1</th>
<th>x1</th>
<th>x2</th>
<th>chr2</th>
<th>y1</th>
<th>y2</th>
<th>color</th>
<th>score</th>
<th>uVarScore</th>
<th>lVarScore</th>
<th>upSign</th>
<th>loSign</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>1</td>
<td>49375000</td>
<td>50805000</td>
<td>1</td>
<td>49375000</td>
<td>50805000</td>
<td>255,255,0</td>
<td>0.8635365988485998</td>
<td>0.3025175910115416</td>
<td>0.2942087461534729</td>
<td>0.4431332556332556</td>
<td>0.6401515151515151</td>
</tr>
<tr class="even">
<td>1</td>
<td>16830000</td>
<td>17230000</td>
<td>1</td>
<td>16830000</td>
<td>17230000</td>
<td>255,255,0</td>
<td>0.20513320424507786</td>
<td>0.3676005813255187</td>
<td>0.22358121923097696</td>
<td>0.4121951219512195</td>
<td>0.4792682926829268</td>
</tr>
<tr class="odd">
<td>1</td>
<td>163355000</td>
<td>164860000</td>
<td>1</td>
<td>163355000</td>
<td>164860000</td>
<td>255,255,0</td>
<td>0.8052794339967723</td>
<td>0.27908025641958345</td>
<td>0.3017420763866275</td>
<td>0.657032586290075</td>
<td>0.5470812683654226</td>
</tr>
<tr class="even">
<td>1</td>
<td>231935000</td>
<td>233400000</td>
<td>1</td>
<td>231935000</td>
<td>233400000</td>
<td>255,255,0</td>
<td>0.43336944864061144</td>
<td>0.28052310457518204</td>
<td>0.27077299003747285</td>
<td>0.451432273589708</td>
<td>0.44314868804664725</td>
</tr>
<tr class="odd">
<td>1</td>
<td>149035000</td>
<td>149430000</td>
<td>1</td>
<td>149035000</td>
<td>149430000</td>
<td>255,255,0</td>
<td>0.8579084086781587</td>
<td>0.3477831683616397</td>
<td>0.18791010793617813</td>
<td>0.45375</td>
<td>0.805625</td>
</tr>
</tbody>
</table>
<p>Explanations of each field are as follows:</p>
<ul>
<li>chromosome = the chromosome that the domain is located on</li>
<li>x1,x2/y1,y2 = the interval spanned by the domain (contact domains manifest as squares on the diagonal of a Hi-C matrix and as such: x1=y1, x2=y2)</li>
<li>color = the color that the feature will be rendered as if loaded in Juicebox</li>
<li>corner_score = the corner score, a score indicating the likelihood that a pixel is at the corner of a contact domain. Higher values indicate a greater likelihood of being at the corner of a domain</li>
<li>Uvar = the variance of the upper triangle</li>
<li>Lvar = the variance of the lower triangle</li>
<li>Usign = -1*(sum of the sign of the entries in the upper triangle)</li>
<li>Lsign = sum of the sign of the entries in the lower triangle</li>
</ul>
<p>For the purposes of this workshop, we are only interested in the first 3 columns, the chromosome, and the start and end coordinates of each called TAD. Since this is performed per chromosome, each chromosome specific file needs to be concatenated to obtain the full list of TADs on the entire genome. For convenience, the TADs called at 5kb resolution for GM12878 using Arrowhead are provided internally in the <em>preciseTAD</em> package.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="st">"arrowhead_gm12878_5kb"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">arrowhead_gm12878_5kb</span><span class="op">)</span>
<span class="co">#&gt;   V1        V2        V3</span>
<span class="co">#&gt; 1  1  49375000  50805000</span>
<span class="co">#&gt; 2  1  16830000  17230000</span>
<span class="co">#&gt; 3  1 163355000 164860000</span>
<span class="co">#&gt; 4  1 231935000 233400000</span>
<span class="co">#&gt; 5  1 149035000 149430000</span>
<span class="co">#&gt; 6  1   3995000   5505000</span></code></pre></div>
</div>
<div id="obtaining-functional-genomic-elements" class="section level1">
<h1 class="hasAnchor">
<a href="#obtaining-functional-genomic-elements" class="anchor"></a>Obtaining functional genomic elements</h1>
<p>The next piece of data that we need are the functional genomic elements used to construct the feature space. These are processed narrowPeak ChIP-seq data in the form of .BED files with chromosome, and start and end coordinates of called peaks of signal enrichment. For this workshop we consider 43 transcription factor binding sites.</p>
<!--
| Genomic Class                               | Element         |
|---------------------------------------------|-----------------|
| BroadHMM chromatin segmentations            | 10TxnElongation |
|                                             | 11WeakTxn       |
|                                             | 12Repressed     |
|                                             | 13Heterochromlo |
|                                             | 14RepetitiveCNV |
|                                             | 15RepetitiveCNV |
|                                             | 1ActivePromoter |
|                                             | 2WeakPromoter   |
|                                             | 3PoisedPromoter |
|                                             | 4StrongEnhancer |
|                                             | 5StrongEnhancer |
|                                             | 6WeakEnhancer   |
|                                             | 7WeakEnhancer   |
|                                             | 8Insulator      |
|                                             | 9TxnTransition  |
| Histone Modifications (HM)                  | H2az            |
|                                             | H3k27ac         |
|                                             | H3k27me3        |
|                                             | H3k36me3        |
|                                             | H3k4me1         |
|                                             | H3k4me2         |
|                                             | H3k4me3         |
|                                             | H3k79me2        |
|                                             | H3k9ac          |
|                                             | H3k9me3         |
|                                             | H4k20me1        |
| Transcription Factor Binding Sites   (TFBS) | Atf3            |
|                                             | Cfos            |
|                                             | Cmyc            |
|                                             | Ctcf            |
|                                             | Egr1            |
|                                             | Ets1            |
|                                             | Gabp            |
|                                             | Jund            |
|                                             | Max             |
|                                             | Mazab85725      |
|                                             | Mef2a           |
|                                             | Mxi1            |
|                                             | P300            |
|                                             | Pmlsc71910      |
|                                             | Pol2            |
|                                             | Pu1             |
|                                             | Rad21           |
|                                             | Rfx5            |
|                                             | Sin3a           |
|                                             | Six5            |
|                                             | Smc3ab9263      |
|                                             | Sp1             |
|                                             | Srf             |
|                                             | Stat1           |
|                                             | Taf1            |
|                                             | Tr4             |
|                                             | Yy1             |
|                                             | Znf143          |
-->
<p>Cell type-specific genomic annotation data can be downloaded from the <a href="http://www.encodeproject.org/chip-seq-matrix/?type=Experiment&amp;replicates.library.biosample.donor.organism.scientific_name=Homo%20sapiens&amp;assay_title=TF%20ChIP-seq&amp;status=released">ENCODE</a> data portal as BED files. For convenience, we have included the list of elements in the <code>\inst\extdata\genomicElements</code> folder of the github repository <a href="https://github.com/dozmorovlab/preciseTADworkshop">here</a>.</p>
</div>
<div id="describing-precisetad-functionality" class="section level1">
<h1 class="hasAnchor">
<a href="#describing-precisetad-functionality" class="anchor"></a>Describing <em>preciseTAD</em> functionality</h1>
<p>The training/testing data used for modeling is represented as a matrix with rows being genomic regions, columns being genomic annotations, and cells containing measures of association between them. Users have the option to concatenate genomic regions from multiple chromosomes.</p>
<p>To create the row-wise dimension of the data, <code>preciseTAD</code> uses <em>shifted binning</em>, a strategy for making the dimensions of the data matrix used for modeling by segmenting the linear genome into nonoverlapping regions. This step is transparent for the user. To create shifted bins, chromosome-specific bins start at half of the resolution <em>r</em>, and continue in congruent intervals of length <em>r</em> until the end of the chromosome (<em>mod r</em> + <em>r/2</em>), using hg19 genomic coordinates. The shifted genomic bins, are then defined as boundary regions (<em>Y = 1</em>) if they contain a called boundary, and non-boundary regions (<em>Y = 0</em>) otherwise, thus establishing the binary response vector (<strong>Y</strong>) used for classification (Figure 1). Intuitively, shifted bins are centered on borders between the original bins, thus capturing potential boundaries.</p>
<center>
<div class="figure">
<img src="figures/response_vector.png" style="width:50.0%" alt=""><p class="caption"><strong>Figure 1. Creating the response vector for binary classification.</strong></p>
</div>
</center>
<!--
<div class="figure" style="text-align: center">
<img src="/__w/preciseTADworkshop/preciseTADworkshop/vignettes/figures/response_vector.png" alt="Figure 1. Creating the response vector for binary classification." width="50%" />
<p class="caption">Figure 1. Creating the response vector for binary classification.</p>
</div>
-->
<p>The column-wise dimension is formed by genomic annotations, such as transcription factor binding sites (TFBSs). In this workshop we focus on the <em>(<span class="math inline">\(log_{2}\)</span>) distances</em>, which enumerate the genomic distance from the center of each genomic bin to the center of the nearest ChIP-seq peak region of interest, although other options are available (see ?preciseTAD::createTADdata). These quantaties form the feature space <span class="math inline">\(\textbf{X} = \{X_{1}, X_{2} \cdots X_{p} \}\)</span> (Figure 2). The customized training/testing data formation offered by <code>preciseTAD</code> allows users to implement any binary classification machine learning algorithm.</p>
<center>
<div class="figure">
<img src="figures/feature_space.png" style="width:50.0%" alt=""><p class="caption"><strong>Figure 2. Creating the feature space for binary classification.</strong></p>
</div>
</center>
<p><code>preciseTAD</code> provides functionality to implement a random forest (RF) model, allowing tuning hyperparameters and applying feature reduction. The primary inputs are the training and testing data, a list of hyperparameter values, the number of cross-validation folds to use (if a grid of hyperparameter values is provided), and the metric used for optimization. The output includes the model object (necessary for downstream prediction of boundaries), the variable importance values, and a list of performance metrics when validating the model on the testing data. This model is then used to predict base-level precise boundary locations.</p>
<p>To predict the base-level location of domain boundaries, the <code>preciseTAD</code> model is applied for each base annotated with the aforementioned genomic annotations. The base-level probabilities are clustered using density-based clustering and scalable data partitioning techniques, to narrow boundary regions and points. First, the probability vector, <span class="math inline">\(p_{n_{i}}\)</span>, is extracted (<span class="math inline">\(n_{i}\)</span> representing the length of chromosome <span class="math inline">\(i\)</span>). Next, <em>DBSCAN</em> (Density-based Spatial Clustering of Applications with Noise) <span class="citation">(Ester et al. 1996; Hahsler, Piekenbrock, and Doran 2019)</span> is applied to the matrix of pairwise genomic distances between bases with <span class="math inline">\(p_{n_{i}} \ge t\)</span>, where <span class="math inline">\(t\)</span> is a threshold determined by the user. The resulting clusters of highly predictive bases identified by DBSCAN are termed <em>preciseTAD boundary regions</em> (PTBR). To precisely identify a single base among each PTBR, <em>preciseTAD</em> implements partitioning around medoids (PAM) within each cluster. The corresponding cluster medoid is defined as a <em>preciseTAD boundary point</em> (PTBP), making it the most representative and biologically meaningful base within each clustered PTBR. The output includes a list of genomic coordinates of PTBPs and PTBSs.</p>
<p><code>preciseTAD</code> allows us to use the pre-trained model to predict the precise location of domain boundaries in cell types without Hi-C data but with genome annotation data. Specifically, only cell-line-specific ChIP-seq data (BED format) for CTCF, RAD21, SMC3, and ZNF143 transcription factor binding sites is required. We provide models pre-trained using GM12878 and K562 genome annotation data, and Arrowhead- and Peakachu-predicted boundaries for chromosome-specific prediction of precise domain boundaries in other cell lines.</p>
</div>
<div id="the-precisetad-r-package" class="section level1">
<h1 class="hasAnchor">
<a href="#the-precisetad-r-package" class="anchor"></a>The <em>preciseTAD</em> R package</h1>
<p>As an example of the functionality of <em>preciseTAD</em>, lets consider the goal of predicting TAD boundary coordinates for CHR14 on GM12878 at 5kb. First, we will build an optimized model to predict TAD boundary <em>regions</em>, then we will leverage said model to predict boundaries on CHR14 at base-level resolution. We will consider the heuristic of training a model on chromosomes 1-13 &amp; 15-22, while reserving data from CHR14 as the test data.</p>
<p>Scripts used to generate the figures in the following sections can be found <a href="https://github.com/dozmorovlab/preciseTADworkshop/tree/main/scripts_for_figures">here</a>.</p>
<div id="model-building" class="section level2">
<h2 class="hasAnchor">
<a href="#model-building" class="anchor"></a>Model building</h2>
<p>Users will need to transform the TAD coordinate data into a GRanges object of unique boundary coordinates using the <code><a href="https://rdrr.io/pkg/preciseTAD/man/extractBoundaries.html">preciseTAD::extractBoundaries()</a></code> function. Here we extract boundaries for CHR1-CHR22 because all chromosomes will be in use (either in the training set or the testing set). We specify <code>preprocess=FALSE</code> because we are only interested in boundaries and not filtering TADs by length. Lastly, we specify <code>resolution=5000</code> to match the resolution used by <em>Arrowhead</em> (although this argument is ignored given that <code>preprocess=FALSE</code>). As shown below, there were a total of 16153 unique TAD boundaries reported by <em>Arrowhead</em> for chromosomes 1-22.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">bounds</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/preciseTAD/man/extractBoundaries.html">extractBoundaries</a></span><span class="op">(</span>domains.mat <span class="op">=</span> <span class="va">arrowhead_gm12878_5kb</span>, 
                            preprocess <span class="op">=</span> <span class="cn">FALSE</span>, 
                            CHR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"CHR"</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span>, 
                            resolution <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span>
<span class="co"># View unique boundaries</span>
<span class="va">bounds</span>
<span class="co">#&gt; GRanges object with 16153 ranges and 0 metadata columns:</span>
<span class="co">#&gt;           seqnames    ranges strand</span>
<span class="co">#&gt;              &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="co">#&gt;       [1]     chr1    815000      *</span>
<span class="co">#&gt;       [2]     chr1    890000      *</span>
<span class="co">#&gt;       [3]     chr1    915000      *</span>
<span class="co">#&gt;       [4]     chr1   1005000      *</span>
<span class="co">#&gt;       [5]     chr1   1015000      *</span>
<span class="co">#&gt;       ...      ...       ...    ...</span>
<span class="co">#&gt;   [16149]    chr22  50815000      *</span>
<span class="co">#&gt;   [16150]    chr22  50935000      *</span>
<span class="co">#&gt;   [16151]    chr22  50945000      *</span>
<span class="co">#&gt;   [16152]    chr22  51060000      *</span>
<span class="co">#&gt;   [16153]    chr22  51235000      *</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 22 sequences from an unspecified genome; no seqlengths</span></code></pre></div>
<p>Additionally, users will need to add each of the respective .BED files,representing the functional genomic elements of choice, to a single <code>GenomicRangesList</code> object. This can be done via the <code><a href="https://rdrr.io/pkg/preciseTAD/man/bedToGRangesList.html">preciseTAD::bedToGRangesList()</a></code> function. The <code>signal</code> argument refers to the column in the BED files containing peak signal strength values and is used to assign metadata to the corresponding GRanges (this needs to be the same column for every .BED file; only necessary for downstream plotting).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">path</span> <span class="op">&lt;-</span> <span class="st">"../inst/extdata/genomicElements"</span>
<span class="va">path</span>
<span class="co">#&gt; [1] "../inst/extdata/genomicElements"</span>
<span class="va">tfbsList</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/preciseTAD/man/bedToGRangesList.html">bedToGRangesList</a></span><span class="op">(</span>filepath<span class="op">=</span><span class="va">path</span>, bedList<span class="op">=</span><span class="cn">NULL</span>, bedNames<span class="op">=</span><span class="cn">NULL</span>, pattern <span class="op">=</span> <span class="st">"*.bed"</span>, signal<span class="op">=</span><span class="fl">4</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">tfbsList</span><span class="op">)</span>
<span class="co">#&gt;  [1] "Gm12878-Bcl3-Haib"          "Gm12878-Bclaf101388-Haib"  </span>
<span class="co">#&gt;  [3] "Gm12878-Bhlh-Sydh"          "Gm12878-Cebpbsc150-Haib"   </span>
<span class="co">#&gt;  [5] "Gm12878-Chd1a301218a-Sydh"  "Gm12878-Chd2ab68301-Sydh"  </span>
<span class="co">#&gt;  [7] "Gm12878-Cmyc-Uta"           "Gm12878-Ctcf-Broad"        </span>
<span class="co">#&gt;  [9] "Gm12878-E2f4-Sydh"          "Gm12878-Egr1-Haib"         </span>
<span class="co">#&gt; [11] "Gm12878-Elf1sc631-Haib"     "Gm12878-Elk112771-Sydh"    </span>
<span class="co">#&gt; [13] "Gm12878-Ets1-Haib"          "Gm12878-Ezh239875-Broad"   </span>
<span class="co">#&gt; [15] "Gm12878-Gabp-Haib"          "Gm12878-Jund-Sydh"         </span>
<span class="co">#&gt; [17] "Gm12878-Max-Sydh"           "Gm12878-Mazab85725-Sydh"   </span>
<span class="co">#&gt; [19] "Gm12878-Mef2a-Haib"         "Gm12878-Mxi1-Sydh"         </span>
<span class="co">#&gt; [21] "Gm12878-Nfyb-Sydh"          "Gm12878-Nrf1-Sydh"         </span>
<span class="co">#&gt; [23] "Gm12878-Nrsf-Haib"          "Gm12878-P300-Sydh"         </span>
<span class="co">#&gt; [25] "Gm12878-Pmlsc71910-Haib"    "Gm12878-Pol2-Haib"         </span>
<span class="co">#&gt; [27] "Gm12878-Pol24h8-Haib"       "Gm12878-Pol2s2-Sydh"       </span>
<span class="co">#&gt; [29] "Gm12878-Pu1-Haib"           "Gm12878-Rad21-Haib"        </span>
<span class="co">#&gt; [31] "Gm12878-Rfx5-Sydh"          "Gm12878-Sin3a-Sydh"        </span>
<span class="co">#&gt; [33] "Gm12878-Six5-Haib"          "Gm12878-Smc3-Sydh"         </span>
<span class="co">#&gt; [35] "Gm12878-Sp1-Haib"           "Gm12878-Srf-Haib"          </span>
<span class="co">#&gt; [37] "Gm12878-Stat5asc74442-Haib" "Gm12878-Taf1-Haib"         </span>
<span class="co">#&gt; [39] "Gm12878-Tblr1ab24550-Sydh"  "Gm12878-Tbp-Sydh"          </span>
<span class="co">#&gt; [41] "Gm12878-Usf1-Haib"          "Gm12878-Usf2-Sydh"         </span>
<span class="co">#&gt; [43] "Gm12878-Znf143-Sydh"</span></code></pre></div>
<p>Using the “ground-truth” boundaries and the annotations, we can build the data matrix that will be used for predictive modeling. The <code><a href="https://rdrr.io/pkg/preciseTAD/man/createTADdata.html">preciseTAD::createTADdata()</a></code> function can create the training and testing data simultaneously as a list object. Here, we specify to train on chromosomes 1-22 and test on chromosome 14. Additionally, we specify <code>resolution = 5000</code> to construct 5kb shifted genomic bins (to match the Hi-C data resolution), <code>featureType = "distance"</code> for a <span class="math inline">\(log_2(distance + 1)\)</span>-type feature space, and <code>resampling = "rus"</code> to apply random under-sampling (RUS) on the training data to balance classes of boundary vs. non-boundary regions. We also specify a seed to ensure reproducibility when performing the resampling. The result is a list containing two data frames: (1) the resampled (if specified) training data, and (2) the testing data.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>
<span class="va">tadData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/preciseTAD/man/createTADdata.html">createTADdata</a></span><span class="op">(</span>bounds.GR <span class="op">=</span> <span class="va">bounds</span>,
                         resolution <span class="op">=</span> <span class="fl">5000</span>,
                         genomicElements.GR <span class="op">=</span> <span class="va">tfbsList</span>,
                         featureType <span class="op">=</span> <span class="st">"distance"</span>,
                         resampling <span class="op">=</span> <span class="st">"rus"</span>,
                         trainCHR <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="st">"CHR"</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span>,<span class="fl">10</span><span class="op">:</span><span class="fl">13</span>,<span class="fl">15</span><span class="op">:</span><span class="fl">22</span><span class="op">)</span><span class="op">)</span>,
                         predictCHR <span class="op">=</span> <span class="st">"CHR14"</span><span class="op">)</span>

<span class="co"># View subset of training data</span>
<span class="va">tadData</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>
<span class="co">#&gt;         y Gm12878-Bcl3-Haib Gm12878-Bclaf101388-Haib Gm12878-Bhlh-Sydh</span>
<span class="co">#&gt; 5799   No          14.25414                 14.24822          14.24652</span>
<span class="co">#&gt; 32930  No          11.80090                 14.88312          14.09119</span>
<span class="co">#&gt; 5767  Yes          11.10656                 10.58871          10.99081</span>
<span class="co">#&gt; 8504  Yes          16.48621                 18.09989          17.95111</span>
<span class="co">#&gt; 39486  No          13.57884                 18.67059          13.57837</span>
<span class="co"># Check it is balanced</span>
<span class="fu"><a href="https://rdrr.io/r/base/table.html">table</a></span><span class="op">(</span><span class="va">tadData</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">y</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;    No   Yes </span>
<span class="co">#&gt; 14913 14913</span>

<span class="co"># View subset of testing data</span>
<span class="va">tadData</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>
<span class="co">#&gt;    y Gm12878-Bcl3-Haib Gm12878-Bclaf101388-Haib Gm12878-Bhlh-Sydh</span>
<span class="co">#&gt; 1 No          24.30718                 24.30793          24.30793</span>
<span class="co">#&gt; 2 No          24.30683                 24.30759          24.30758</span>
<span class="co">#&gt; 3 No          24.30649                 24.30724          24.30723</span>
<span class="co">#&gt; 4 No          24.30614                 24.30689          24.30688</span>
<span class="co">#&gt; 5 No          24.30579                 24.30654          24.30654</span></code></pre></div>
</div>
<div id="optimization-and-model-evaluation" class="section level2">
<h2 class="hasAnchor">
<a href="#optimization-and-model-evaluation" class="anchor"></a>Optimization and model evaluation</h2>
<p>We can now implement our machine learning algorithm of choice to predict boundary regions. Here, we opt for the random forest algorithm for binary classification. <code>preciseTAD</code> offers functionality for performing recursive feature elimination (RFE) as a form of feature reduction through the use of the <code><a href="https://rdrr.io/pkg/preciseTAD/man/TADrfe.html">preciseTAD::TADrfe()</a></code> function, which is a wrapper for the <code>rfe</code> function in the <code>caret</code> package <span class="citation">(Kuhn 2012)</span>. <code><a href="https://rdrr.io/pkg/preciseTAD/man/TADrfe.html">preciseTAD::TADrfe()</a></code> implements a random forest model on the best subset of features from 2 to the maximum number of features in the data by powers of 2, using 5-fold cross-validation. We specify accuracy as the performance metric.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span>
<span class="va">rfe_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/preciseTAD/man/TADrfe.html">TADrfe</a></span><span class="op">(</span>trainData <span class="op">=</span> <span class="va">tadData</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,
                 tuneParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>ntree <span class="op">=</span> <span class="fl">500</span>, nodesize <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
                 cvFolds <span class="op">=</span> <span class="fl">5</span>,
                 cvMetric <span class="op">=</span> <span class="st">"Accuracy"</span>,
                 verbose <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Results from RFE indicate that performance levels off when only considering the top 8 most predictive transcription factors (Figure 3A). We can further optimize our model by evaluating the predictive importance of these top 8 genomic elements across each cross-fold via a heatmap. After aggregating the variable importance values across each cross-fold, we see that SMC3, RAD21, CTCF, and ZNF143 stand out as the most predictive elements TAD-boundary regions (Figure 3B). These are known components of the loop-extrusion model that has been proposed as a mechanism for the 3D architecture of the human genome <span class="citation">(Sanborn et al. 2015; Fudenberg et al. 2016; Hansen et al. 2018)</span>.</p>
<center>
<div class="figure">
<img src="figures/rfe_results.png" style="width:75.0%" alt=""><p class="caption"><strong>Figure 3. Optimizing the TAD boundary region prediction model.</strong> (A) Recursive feature elimination. (B) Variable importance heatmap across each of the 5 cross-validation folds.</p>
</div>
</center>
<p>Now that we have suitably reduced our feature space, we can implement a random forest algorithm built simply on the TFBS mentioned above (SMC3, RAD21, CTCF, and ZNF143). We can take advantage of the <code><a href="https://rdrr.io/pkg/preciseTAD/man/TADrandomForest.html">preciseTAD::TADrandomForest()</a></code> function, which is a wrapper of the <code>randomForest</code> package <span class="citation">(Breiman 2001; Liaw, Wiener, and others 2002)</span>. We specify the training and testing data, the hyperparameter values, the number of cross-validation folds, the performance metric to consider (here, accuracy), the seed to initialize for reproducibility, an indicator for retaining the model object, an indicator for retaining variable importances, the variable importance measure to consider (here, mean decrease in accuracy (MDA)), and an indicator for retaining model performances based on the test data. The function returns a list containing: 1) a trained model from caret with model information, 2) a <code>data.frame</code> of variable importance for each feature included in the model, and 3) a <code>data.frame</code> of various model performance metrics. Model performances are shown in Table 1.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Restrict the data matrix to include only SMC3, RAD21, CTCF, and ZNF143</span>
<span class="va">tadDataFiltTrain</span> <span class="op">&lt;-</span> <span class="va">tadData</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Ctcf"</span>,<span class="st">"Rad21"</span>,<span class="st">"Smc3"</span>,<span class="st">"Znf143"</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">"|"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">tadData</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>
<span class="va">tadDataFiltTest</span> <span class="op">&lt;-</span> <span class="va">tadData</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html">grep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Ctcf"</span>,<span class="st">"Rad21"</span>,<span class="st">"Smc3"</span>,<span class="st">"Znf143"</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">"|"</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">tadData</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">]</span>

<span class="co"># Run RF</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">123</span><span class="op">)</span>
<span class="va">tadModel</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/preciseTAD/man/TADrandomForest.html">TADrandomForest</a></span><span class="op">(</span>trainData  <span class="op">=</span> <span class="va">tadDataFiltTrain</span>,
                            testData   <span class="op">=</span> <span class="va">tadDataFiltTest</span>,
                            tuneParams <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>mtry     <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>,<span class="fl">3</span><span class="op">)</span>,
                                              ntree    <span class="op">=</span> <span class="fl">500</span>,
                                              nodesize <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,
                            cvFolds      <span class="op">=</span> <span class="fl">3</span>,
                            cvMetric     <span class="op">=</span> <span class="st">"Accuracy"</span>,
                            verbose      <span class="op">=</span> <span class="cn">TRUE</span>,
                            model        <span class="op">=</span> <span class="cn">TRUE</span>,
                            importances  <span class="op">=</span> <span class="cn">TRUE</span>,
                            impMeasure   <span class="op">=</span> <span class="st">"MDA"</span>,
                            performances <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>
<span class="co">#&gt; + Fold1: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; - Fold1: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; + Fold1: mtry=3, ntree=500, nodesize=1 </span>
<span class="co">#&gt; - Fold1: mtry=3, ntree=500, nodesize=1 </span>
<span class="co">#&gt; + Fold2: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; - Fold2: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; + Fold2: mtry=3, ntree=500, nodesize=1 </span>
<span class="co">#&gt; - Fold2: mtry=3, ntree=500, nodesize=1 </span>
<span class="co">#&gt; + Fold3: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; - Fold3: mtry=2, ntree=500, nodesize=1 </span>
<span class="co">#&gt; + Fold3: mtry=3, ntree=500, nodesize=1 </span>
<span class="co">#&gt; - Fold3: mtry=3, ntree=500, nodesize=1 </span>
<span class="co">#&gt; Aggregating results</span>
<span class="co">#&gt; Selecting tuning parameters</span>
<span class="co">#&gt; Fitting mtry = 2, ntree = 500, nodesize = 1 on full training set</span>

<span class="co"># View model performances</span>
<span class="va">performances</span> <span class="op">&lt;-</span> <span class="va">tadModel</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>
<span class="va">performances</span><span class="op">$</span><span class="va">Performance</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html">round</a></span><span class="op">(</span><span class="va">performances</span><span class="op">$</span><span class="va">Performance</span>, digits <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html">rownames</a></span><span class="op">(</span><span class="va">performances</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="cn">NULL</span> <span class="co">#performances$Metric</span>
<span class="fu"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span><span class="op">(</span><span class="va">performances</span>, caption <span class="op">=</span> <span class="st">"Table 1. List of model performances."</span><span class="op">)</span></code></pre></div>
<table class="table">
<caption>Table 1. List of model performances.</caption>
<thead><tr class="header">
<th align="left">Metric</th>
<th align="right">Performance</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">TN</td>
<td align="right">14835.00</td>
</tr>
<tr class="even">
<td align="left">FN</td>
<td align="right">132.00</td>
</tr>
<tr class="odd">
<td align="left">FP</td>
<td align="right">5479.00</td>
</tr>
<tr class="even">
<td align="left">TP</td>
<td align="right">423.00</td>
</tr>
<tr class="odd">
<td align="left">Total</td>
<td align="right">20869.00</td>
</tr>
<tr class="even">
<td align="left">Sensitivity</td>
<td align="right">0.76</td>
</tr>
<tr class="odd">
<td align="left">Specificity</td>
<td align="right">0.73</td>
</tr>
<tr class="even">
<td align="left">Kappa</td>
<td align="right">0.09</td>
</tr>
<tr class="odd">
<td align="left">Accuracy</td>
<td align="right">0.73</td>
</tr>
<tr class="even">
<td align="left">BalancedAccuracy</td>
<td align="right">0.75</td>
</tr>
<tr class="odd">
<td align="left">Precision</td>
<td align="right">0.07</td>
</tr>
<tr class="even">
<td align="left">FPR</td>
<td align="right">0.27</td>
</tr>
<tr class="odd">
<td align="left">FNR</td>
<td align="right">0.01</td>
</tr>
<tr class="even">
<td align="left">NPV</td>
<td align="right">0.99</td>
</tr>
<tr class="odd">
<td align="left">MCC</td>
<td align="right">0.18</td>
</tr>
<tr class="even">
<td align="left">F1</td>
<td align="right">0.13</td>
</tr>
<tr class="odd">
<td align="left">AUC</td>
<td align="right">0.82</td>
</tr>
<tr class="even">
<td align="left">Youden</td>
<td align="right">0.49</td>
</tr>
<tr class="odd">
<td align="left">AUPRC</td>
<td align="right">0.09</td>
</tr>
</tbody>
</table>
</div>
<div id="reading-in-model-objects-using-precisetadhub-must-be-using-r-verion-4-1-or-r-devel" class="section level2">
<h2 class="hasAnchor">
<a href="#reading-in-model-objects-using-precisetadhub-must-be-using-r-verion-4-1-or-r-devel" class="anchor"></a>Reading in model objects using <code>preciseTADhub</code> (must be using R verion &gt;= 4.1 or R-devel)</h2>
<p>For convenience, we have provided an ExperimentData package to supplement <code>preciseTAD</code>, called <code>preciseTADhub</code>. <code>preciseTADhub</code> offers users access to pre-trained random forest classification models built on CTCF, RAD21, SMC3, and ZNF143. Each of the 84 (2 cell lines <span class="math inline">\(\times\)</span> 2 ground truth boundaries <span class="math inline">\(\times\)</span> 21 autosomal chromosomes) files (.RDS) are stored as lists containing two objects: 1) a train object from  with RF model information, and 2) a data.frame of variable importance for each genomic annotation included in the model. The file names are structured as follows:</p>
<p><span class="math display">\[i\_j\_k\_l.rds\]</span></p>
<p>where <span class="math inline">\(i\)</span> denotes the chromosome that was used as a holdout {CHR1, CHR2, …, CHR21, CHR22} (i.e. for testing; meaning all other chromosomes were used for training), <span class="math inline">\(j\)</span> denotes the cell line {GM12878, K562}, <span class="math inline">\(k\)</span> denotes the resolution (size of genomic bins) {5kb, 10kb}, and <span class="math inline">\(l\)</span> denotes the TAD/loop caller used to define ground truth {Arrowhead, Peakachu}.</p>
<p>Table 2 shows the file names and the corresponding ExperimentHub (EH) IDs. Since we want to make TAD boundary predictions on CHR14 for GM12878, we opt to read in the “CHR14_GM12878_5kb_Arrowhead.rds” file. This corresponds to the EH3863 EHID.</p>
<table class="table">
<thead><tr class="header">
<th>FileName</th>
<th>EHID</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>CHR1_GM12878_5kb_Arrowhead.rds</td>
<td>EH3815</td>
</tr>
<tr class="even">
<td>CHR1_GM12878_10kb_Peakachu.rds</td>
<td>EH3816</td>
</tr>
<tr class="odd">
<td>CHR1_K562_5kb_Arrowhead.rds</td>
<td>EH3817</td>
</tr>
<tr class="even">
<td>CHR1_K562_10kb_Peakachu.rds</td>
<td>EH3818</td>
</tr>
<tr class="odd">
<td>CHR2_GM12878_5kb_Arrowhead.rds</td>
<td>EH3819</td>
</tr>
<tr class="even">
<td>CHR2_GM12878_10kb_Peakachu.rds</td>
<td>EH3820</td>
</tr>
<tr class="odd">
<td>CHR2_K562_5kb_Arrowhead.rds</td>
<td>EH3821</td>
</tr>
<tr class="even">
<td>CHR2_K562_10kb_Peakachu.rds</td>
<td>EH3822</td>
</tr>
<tr class="odd">
<td>CHR3_GM12878_5kb_Arrowhead.rds</td>
<td>EH3823</td>
</tr>
<tr class="even">
<td>CHR3_GM12878_10kb_Peakachu.rds</td>
<td>EH3824</td>
</tr>
<tr class="odd">
<td>CHR3_K562_5kb_Arrowhead.rds</td>
<td>EH3825</td>
</tr>
<tr class="even">
<td>CHR3_K562_10kb_Peakachu.rds</td>
<td>EH3826</td>
</tr>
<tr class="odd">
<td>CHR4_GM12878_5kb_Arrowhead.rds</td>
<td>EH3827</td>
</tr>
<tr class="even">
<td>CHR4_GM12878_10kb_Peakachu.rds</td>
<td>EH3828</td>
</tr>
<tr class="odd">
<td>CHR4_K562_5kb_Arrowhead.rds</td>
<td>EH3829</td>
</tr>
<tr class="even">
<td>CHR4_K562_10kb_Peakachu.rds</td>
<td>EH3830</td>
</tr>
<tr class="odd">
<td>CHR5_GM12878_5kb_Arrowhead.rds</td>
<td>EH3831</td>
</tr>
<tr class="even">
<td>CHR5_GM12878_10kb_Peakachu.rds</td>
<td>EH3832</td>
</tr>
<tr class="odd">
<td>CHR5_K562_5kb_Arrowhead.rds</td>
<td>EH3833</td>
</tr>
<tr class="even">
<td>CHR5_K562_10kb_Peakachu.rds</td>
<td>EH3834</td>
</tr>
<tr class="odd">
<td>CHR6_GM12878_5kb_Arrowhead.rds</td>
<td>EH3835</td>
</tr>
<tr class="even">
<td>CHR6_GM12878_10kb_Peakachu.rds</td>
<td>EH3836</td>
</tr>
<tr class="odd">
<td>CHR6_K562_5kb_Arrowhead.rds</td>
<td>EH3837</td>
</tr>
<tr class="even">
<td>CHR6_K562_10kb_Peakachu.rds</td>
<td>EH3838</td>
</tr>
<tr class="odd">
<td>CHR7_GM12878_5kb_Arrowhead.rds</td>
<td>EH3839</td>
</tr>
<tr class="even">
<td>CHR7_GM12878_10kb_Peakachu.rds</td>
<td>EH3840</td>
</tr>
<tr class="odd">
<td>CHR7_K562_5kb_Arrowhead.rds</td>
<td>EH3841</td>
</tr>
<tr class="even">
<td>CHR7_K562_10kb_Peakachu.rds</td>
<td>EH3842</td>
</tr>
<tr class="odd">
<td>CHR8_GM12878_5kb_Arrowhead.rds</td>
<td>EH3843</td>
</tr>
<tr class="even">
<td>CHR8_GM12878_10kb_Peakachu.rds</td>
<td>EH3844</td>
</tr>
<tr class="odd">
<td>CHR8_K562_5kb_Arrowhead.rds</td>
<td>EH3845</td>
</tr>
<tr class="even">
<td>CHR8_K562_10kb_Peakachu.rds</td>
<td>EH3846</td>
</tr>
<tr class="odd">
<td>CHR10_GM12878_5kb_Arrowhead.rds</td>
<td>EH3847</td>
</tr>
<tr class="even">
<td>CHR10_GM12878_10kb_Peakachu.rds</td>
<td>EH3848</td>
</tr>
<tr class="odd">
<td>CHR10_K562_5kb_Arrowhead.rds</td>
<td>EH3849</td>
</tr>
<tr class="even">
<td>CHR10_K562_10kb_Peakachu.rds</td>
<td>EH3850</td>
</tr>
<tr class="odd">
<td>CHR11_GM12878_5kb_Arrowhead.rds</td>
<td>EH3851</td>
</tr>
<tr class="even">
<td>CHR11_GM12878_10kb_Peakachu.rds</td>
<td>EH3852</td>
</tr>
<tr class="odd">
<td>CHR11_K562_5kb_Arrowhead.rds</td>
<td>EH3853</td>
</tr>
<tr class="even">
<td>CHR11_K562_10kb_Peakachu.rds</td>
<td>EH3854</td>
</tr>
<tr class="odd">
<td>CHR12_GM12878_5kb_Arrowhead.rds</td>
<td>EH3855</td>
</tr>
<tr class="even">
<td>CHR12_GM12878_10kb_Peakachu.rds</td>
<td>EH3856</td>
</tr>
<tr class="odd">
<td>CHR12_K562_5kb_Arrowhead.rds</td>
<td>EH3857</td>
</tr>
<tr class="even">
<td>CHR12_K562_10kb_Peakachu.rds</td>
<td>EH3858</td>
</tr>
<tr class="odd">
<td>CHR13_GM12878_5kb_Arrowhead.rds</td>
<td>EH3859</td>
</tr>
<tr class="even">
<td>CHR13_GM12878_10kb_Peakachu.rds</td>
<td>EH3860</td>
</tr>
<tr class="odd">
<td>CHR13_K562_5kb_Arrowhead.rds</td>
<td>EH3861</td>
</tr>
<tr class="even">
<td>CHR13_K562_10kb_Peakachu.rds</td>
<td>EH3862</td>
</tr>
<tr class="odd">
<td>CHR14_GM12878_5kb_Arrowhead.rds</td>
<td>EH3863</td>
</tr>
<tr class="even">
<td>CHR14_GM12878_10kb_Peakachu.rds</td>
<td>EH3864</td>
</tr>
<tr class="odd">
<td>CHR14_K562_5kb_Arrowhead.rds</td>
<td>EH3865</td>
</tr>
<tr class="even">
<td>CHR14_K562_10kb_Peakachu.rds</td>
<td>EH3866</td>
</tr>
<tr class="odd">
<td>CHR15_GM12878_5kb_Arrowhead.rds</td>
<td>EH3867</td>
</tr>
<tr class="even">
<td>CHR15_GM12878_10kb_Peakachu.rds</td>
<td>EH3868</td>
</tr>
<tr class="odd">
<td>CHR15_K562_5kb_Arrowhead.rds</td>
<td>EH3869</td>
</tr>
<tr class="even">
<td>CHR15_K562_10kb_Peakachu.rds</td>
<td>EH3870</td>
</tr>
<tr class="odd">
<td>CHR16_GM12878_5kb_Arrowhead.rds</td>
<td>EH3871</td>
</tr>
<tr class="even">
<td>CHR16_GM12878_10kb_Peakachu.rds</td>
<td>EH3872</td>
</tr>
<tr class="odd">
<td>CHR16_K562_5kb_Arrowhead.rds</td>
<td>EH3873</td>
</tr>
<tr class="even">
<td>CHR16_K562_10kb_Peakachu.rds</td>
<td>EH3874</td>
</tr>
<tr class="odd">
<td>CHR17_GM12878_5kb_Arrowhead.rds</td>
<td>EH3875</td>
</tr>
<tr class="even">
<td>CHR17_GM12878_10kb_Peakachu.rds</td>
<td>EH3876</td>
</tr>
<tr class="odd">
<td>CHR17_K562_5kb_Arrowhead.rds</td>
<td>EH3877</td>
</tr>
<tr class="even">
<td>CHR17_K562_10kb_Peakachu.rds</td>
<td>EH3878</td>
</tr>
<tr class="odd">
<td>CHR18_GM12878_5kb_Arrowhead.rds</td>
<td>EH3879</td>
</tr>
<tr class="even">
<td>CHR18_GM12878_10kb_Peakachu.rds</td>
<td>EH3880</td>
</tr>
<tr class="odd">
<td>CHR18_K562_5kb_Arrowhead.rds</td>
<td>EH3881</td>
</tr>
<tr class="even">
<td>CHR18_K562_10kb_Peakachu.rds</td>
<td>EH3882</td>
</tr>
<tr class="odd">
<td>CHR19_GM12878_5kb_Arrowhead.rds</td>
<td>EH3883</td>
</tr>
<tr class="even">
<td>CHR19_GM12878_10kb_Peakachu.rds</td>
<td>EH3884</td>
</tr>
<tr class="odd">
<td>CHR19_K562_5kb_Arrowhead.rds</td>
<td>EH3885</td>
</tr>
<tr class="even">
<td>CHR19_K562_10kb_Peakachu.rds</td>
<td>EH3886</td>
</tr>
<tr class="odd">
<td>CHR20_GM12878_5kb_Arrowhead.rds</td>
<td>EH3887</td>
</tr>
<tr class="even">
<td>CHR20_GM12878_10kb_Peakachu.rds</td>
<td>EH3888</td>
</tr>
<tr class="odd">
<td>CHR20_K562_5kb_Arrowhead.rds</td>
<td>EH3889</td>
</tr>
<tr class="even">
<td>CHR20_K562_10kb_Peakachu.rds</td>
<td>EH3890</td>
</tr>
<tr class="odd">
<td>CHR21_GM12878_5kb_Arrowhead.rds</td>
<td>EH3891</td>
</tr>
<tr class="even">
<td>CHR21_GM12878_10kb_Peakachu.rds</td>
<td>EH3892</td>
</tr>
<tr class="odd">
<td>CHR21_K562_5kb_Arrowhead.rds</td>
<td>EH3893</td>
</tr>
<tr class="even">
<td>CHR21_K562_10kb_Peakachu.rds</td>
<td>EH3894</td>
</tr>
<tr class="odd">
<td>CHR22_GM12878_5kb_Arrowhead.rds</td>
<td>EH3895</td>
</tr>
<tr class="even">
<td>CHR22_GM12878_10kb_Peakachu.rds</td>
<td>EH3896</td>
</tr>
<tr class="odd">
<td>CHR22_K562_5kb_Arrowhead.rds</td>
<td>EH3897</td>
</tr>
<tr class="even">
<td>CHR22_K562_10kb_Peakachu.rds</td>
<td>EH3898</td>
</tr>
</tbody>
</table>
<p>Table 2: File names and corresponding ExperimentHub (EH) IDs for all .RDS files stored in <code>preciseTADhub</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#Initialize ExperimentHub</span>
<span class="va">hub</span> <span class="op">&lt;-</span> <span class="fu">ExperimentHub</span><span class="op">(</span><span class="op">)</span>
<span class="fu">query</span><span class="op">(</span><span class="va">hub</span>, <span class="st">"preciseTADhub"</span><span class="op">)</span>
<span class="va">myfiles</span> <span class="op">&lt;-</span> <span class="fu">query</span><span class="op">(</span><span class="va">hub</span>, <span class="st">"preciseTADhub"</span><span class="op">)</span>

<span class="va">id</span> <span class="op">&lt;-</span> <span class="st">"EH3863"</span>
<span class="va">tadModel</span> <span class="op">&lt;-</span> <span class="va">myfiles</span><span class="op">[[</span><span class="va">id</span><span class="op">]</span><span class="op">]</span> </code></pre></div>
</div>
<div id="boundary-prediction-with-precisetad" class="section level2">
<h2 class="hasAnchor">
<a href="#boundary-prediction-with-precisetad" class="anchor"></a>Boundary prediction with <em>preciseTAD</em>
</h2>
<p>Recall that our model classifies boundary <strong>regions</strong>, in that each prediction refers to a genomic bin of width 5000 bases. To predict boundary coordinates at the base resolution more precisely, we can leverage our model through the use of the <code><a href="https://rdrr.io/pkg/preciseTAD/man/preciseTAD.html">preciseTAD::preciseTAD()</a></code> function. Conceptually, instead of genomic bins, we annotate each base with the selected genomic annotations and featureType. We then apply our model on this annotation matrix to predict the probability of each base as a boundary given the associated genomic annotations. See the preciseTAD paper for a concise description of the algorithm and the pseudocode.</p>
<p>Suppose we want to precisely predict the domain boundary coordinates for the 1.3Mb section of CHR14:50,070,000-50,800,000. To do so, we specify <code>chromCoords = list(50070000,50800000)</code>. Additionally, we set a threshold of 1.0 for constructing PTBRs using <code>threshold = 1.0</code>. For DBSCAN, we assign 10000 and 3 for the <span class="math inline">\(\epsilon\)</span> and <em>MinPts</em> parameters, respectively. The results is a list containing 3 elements including: 1) the genomic coordinates spanning each preciseTAD predicted region (PTBR), 2) the genomic coordinates of preciseTAD predicted boundaries points (PTBP), and 3) a named list including summary statistics of the following: PTBRWidth - PTBR width, PTBRCoverage - the proportion of bases within a PTBR with probabilities that equal to or exceed the threshold (t=1 by default), DistanceBetweenPTBR - the genomic distance between the end of the previous PTBR and the start of the subsequent PTBR, NumSubRegions - the number of the subregions in each PTBR cluster, SubRegionWidth - the width of the subregion forming each PTBR, DistBetweenSubRegions - the genomic distance between the end of the previous PTBR-specific subregion and the start of the subsequent PTBR-specific subregion, and the normalized enrichment of the genomic annotations used in the model around flanked PTBPs (Figure 4).</p>
<center>
<div class="figure">
<img src="figures/pt_summaries.png" style="width:75.0%" alt=""><p class="caption"><strong>Figure 4. <em>preciseTAD</em> summaries.</strong></p>
</div>
</center>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Restrict to include only SMC3, RAD21, CTCF, and ZNF143</span>
<span class="va">tfbsList_filt</span> <span class="op">&lt;-</span> <span class="va">tfbsList</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">tfbsList</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"Gm12878-Ctcf-Broad"</span>, 
                                            <span class="st">"Gm12878-Rad21-Haib"</span>,
                                            <span class="st">"Gm12878-Smc3-Sydh"</span>,
                                            <span class="st">"Gm12878-Znf143-Sydh"</span><span class="op">)</span><span class="op">]</span>

<span class="co"># Run preciseTAD</span>
<span class="va">pt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/preciseTAD/man/preciseTAD.html">preciseTAD</a></span><span class="op">(</span>genomicElements.GR <span class="op">=</span> <span class="va">tfbsList_filt</span>,
                featureType         <span class="op">=</span> <span class="st">"distance"</span>,
                CHR                 <span class="op">=</span> <span class="st">"CHR14"</span>,
                chromCoords         <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fl">50070000</span>,<span class="fl">51360000</span><span class="op">)</span>,
                tadModel            <span class="op">=</span> <span class="va">tadModel</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>,
                threshold           <span class="op">=</span> <span class="fl">1.0</span>,
                verbose             <span class="op">=</span> <span class="cn">TRUE</span>,
                parallel            <span class="op">=</span> <span class="cn">NULL</span>,
                DBSCAN_params       <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span><span class="fl">10000</span>, <span class="fl">3</span><span class="op">)</span>,
                flank               <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span>
<span class="co">#&gt; [1] "Establishing bp resolution test data using a distance type feature space"</span>
<span class="co">#&gt; [1] "Establishing probability vector"</span>
<span class="co">#&gt; [1] "preciseTAD identified a total of 2918 base pairs whose predictive probability was equal to or exceeded a threshold of 1"</span>
<span class="co">#&gt; [1] "Initializing DBSCAN"</span>
<span class="co">#&gt; [1] "preciseTAD identified 8 PTBRs"</span>
<span class="co">#&gt; [1] "Establishing PTBPs"</span>
<span class="co">#&gt; [1] "Cluster 1 out of 8"</span>
<span class="co">#&gt; [1] "Cluster 2 out of 8"</span>
<span class="co">#&gt; [1] "Cluster 3 out of 8"</span>
<span class="co">#&gt; [1] "Cluster 4 out of 8"</span>
<span class="co">#&gt; [1] "Cluster 5 out of 8"</span>
<span class="co">#&gt; [1] "Cluster 6 out of 8"</span>
<span class="co">#&gt; [1] "Cluster 7 out of 8"</span>
<span class="co">#&gt; [1] "Cluster 8 out of 8"</span>

<span class="co"># View the results</span>
<span class="va">pt</span>
<span class="co">#&gt; $PTBR</span>
<span class="co">#&gt; GRanges object with 8 ranges and 0 metadata columns:</span>
<span class="co">#&gt;       seqnames            ranges strand</span>
<span class="co">#&gt;          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="co">#&gt;   [1]    chr14 50101047-50101136      *</span>
<span class="co">#&gt;   [2]    chr14 50312215-50334062      *</span>
<span class="co">#&gt;   [3]    chr14 50563705-50583007      *</span>
<span class="co">#&gt;   [4]    chr14 50784355-50796085      *</span>
<span class="co">#&gt;   [5]    chr14 50997740-51001756      *</span>
<span class="co">#&gt;   [6]    chr14 51059855-51093997      *</span>
<span class="co">#&gt;   [7]    chr14 51231563-51247722      *</span>
<span class="co">#&gt;   [8]    chr14 51319223-51334404      *</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $PTBP</span>
<span class="co">#&gt; GRanges object with 8 ranges and 0 metadata columns:</span>
<span class="co">#&gt;       seqnames    ranges strand</span>
<span class="co">#&gt;          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="co">#&gt;   [1]    chr14  50101086      *</span>
<span class="co">#&gt;   [2]    chr14  50328921      *</span>
<span class="co">#&gt;   [3]    chr14  50571870      *</span>
<span class="co">#&gt;   [4]    chr14  50787930      *</span>
<span class="co">#&gt;   [5]    chr14  50999455      *</span>
<span class="co">#&gt;   [6]    chr14  51066958      *</span>
<span class="co">#&gt;   [7]    chr14  51241489      *</span>
<span class="co">#&gt;   [8]    chr14  51327392      *</span>
<span class="co">#&gt;   -------</span>
<span class="co">#&gt;   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Summaries</span>
<span class="co">#&gt; $Summaries$PTBRWidth</span>
<span class="co">#&gt;   min   max median      iqr     mean       sd</span>
<span class="co">#&gt; 1  90 34143  15671 10136.75 15309.25 10597.17</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Summaries$PTBRCoverage</span>
<span class="co">#&gt;          min       max    median       iqr      mean        sd</span>
<span class="co">#&gt; 1 0.01098322 0.8666667 0.0278572 0.0212158 0.1339193 0.2964553</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Summaries$DistanceBetweenPTBR</span>
<span class="co">#&gt;     min    max median      iqr     mean       sd</span>
<span class="co">#&gt; 1 58099 229643 201348 101833.5 158698.7 70251.13</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Summaries$NumSubRegions</span>
<span class="co">#&gt;   min max median  iqr   mean       sd</span>
<span class="co">#&gt; 1   2  50     32 8.75 30.125 13.98405</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Summaries$SubRegionWidth</span>
<span class="co">#&gt;   min max median iqr     mean      sd</span>
<span class="co">#&gt; 1   1 193      4   7 12.11203 30.7296</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Summaries$DistBetweenSubRegions</span>
<span class="co">#&gt;   min  max median iqr     mean       sd</span>
<span class="co">#&gt; 1   2 7529     40 389 514.1116 1149.644</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $Summaries$NormilizedEnrichment</span>
<span class="co">#&gt;  Gm12878-Ctcf-Broad  Gm12878-Rad21-Haib   Gm12878-Smc3-Sydh Gm12878-Znf143-Sydh </span>
<span class="co">#&gt;                1.25                1.25                1.25                1.25</span></code></pre></div>
</div>
<div id="using-precisetad-with-juicebox" class="section level2">
<h2 class="hasAnchor">
<a href="#using-precisetad-with-juicebox" class="anchor"></a>Using <em>preciseTAD</em> with Juicebox</h2>
<p>Juicebox is an interface provided by Aiden Lab that allows for superimposing boundary coordinates onto Hi-C contact maps. You can download a desktop version of the application <a href="https://github.com/aidenlab/Juicebox/wiki/Download">here</a>, or use the online version <a href="https://www.aidenlab.org/juicebox">here</a>. To visualize domains demarcated by the predicted boundaries, you must first select a Hi-C map. Using the desktop version, you can import the contact matrix for GM12878 derived by Rao et al. 2014 by choosing <code>File -&gt; Open... -&gt; Rao and Huntley et al. -&gt; GM12878 -&gt; in situ Mbol -&gt; primary</code>. To format <code>preciseTAD</code> results to use in Juicebox, users can take advantage of <code><a href="https://rdrr.io/pkg/preciseTAD/man/juicer_func.html">preciseTAD::juicer_func()</a></code>, which transforms a GRanges object into a data frame as shown below.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Transform</span>
<span class="va">pt_juice</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/preciseTAD/man/juicer_func.html">juicer_func</a></span><span class="op">(</span><span class="va">pt</span><span class="op">$</span><span class="va">PTBP</span><span class="op">)</span></code></pre></div>
<p>You will then need to save the PTBPs as a BED file using <code>write.table</code> as shown below. Once saved, import them into Juicebox using <code>Show Annotation Panel -&gt; 2D Annotations -&gt; Add Local -&gt; pt_juice.bed</code>. An example of using Juicebox to plot 3D domain boundary coordinates is shown in Figure 5.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">filepath</span> <span class="op">=</span> <span class="st">"~/path/to/store/ptbps"</span>
<span class="fu"><a href="https://rdrr.io/r/utils/write.table.html">write.table</a></span><span class="op">(</span><span class="va">pt_juice</span>, 
            <span class="fu"><a href="https://rdrr.io/r/base/file.path.html">file.path</a></span><span class="op">(</span><span class="va">filepath</span>, <span class="st">"pt_juice.bed"</span><span class="op">)</span>,
            quote <span class="op">=</span> <span class="cn">FALSE</span>,
            col.names <span class="op">=</span> <span class="cn">TRUE</span>,
            row.names <span class="op">=</span> <span class="cn">FALSE</span>,
            sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></code></pre></div>
<center>
<div class="figure">
<img src="figures/juicer_screenshot2.png" style="width:75.0%" alt=""><p class="caption"><strong>Figure 5. Juicebox visualization of TAD boundaries called by Arrowhead (blue) vs predicted by <em>preciseTAD</em> (green).</strong> The region presented are the genomic coordinates spanning 50,055,000-51,360,000 on chromosome 14 of GM12878.</p>
</div>
</center>
</div>
<div id="techniques-for-visualizing-tad-boundaries" class="section level2">
<h2 class="hasAnchor">
<a href="#techniques-for-visualizing-tad-boundaries" class="anchor"></a>Techniques for visualizing TAD boundaries</h2>
<p>The biological significance of TAD boundaries can be assessed by evaluating their association with the signal of known molecular drivers of 3D chromatin, namely CTCF, RAD21, SMC3, and ZNF143. Two popular visualization techniques include <em>enriched heatmaps</em> and <em>profile plots</em>, both of which can be accomplished with deepTools <span class="citation">(Ramirez et al. 2016)</span>. Briefly, for an enriched heatmap, a matrix <span class="math inline">\(M_{r \times c}\)</span>, is created where the rows, <span class="math inline">\(r\)</span>, are given by the number of boundaries, either called or predicted. The column dimension, <span class="math inline">\(c\)</span>, is created by flanking each boundary by a given amount (5 kb). The flanking is then broken up into 50 bp segments, for 100 windows on both sides of a given boundary, a total of 200 columns. The cells of the matrix are calculated as a mean coverage value for each window with respect to the signal from the respective ChIP-seq annotation via bigwig files. For a profile plot, the matrix is then summarized by row-wise averages and plotted as a density curve, where the center represents the boundary, and the curve represents the average ChIP-seq peak signal around a flanked region. You can read more about each technique <a href="https://deeptools.readthedocs.io/en/develop/content/list_of_tools.html">here</a>.</p>
<p>Furthermore, we can assess the conservation of TAD boundaries by visualizing the amount of overlap between flanked boundaries across cell lines. Overlaps can be quantified using the Jaccard index defined as</p>
<p><span class="math display">\[J_{(A,B)} = \dfrac{A \cap B}{A \cup B}\]</span></p>
<p>where A and B represent genomic regions created by flanked called and predicted boundaries. That is, between cell lines, the number of overlapping flanked boundaries divided by the total number of flanked boundaries. Wilcoxon Rank-Sum tests can then be used to compare chromosome-specific Jaccard indices across cell lines.</p>
<p>Below is an example script of how to create an enriched heatmap and profile plot for a set of boundaries using deepTools (version 2.0). Users will first need to create the matrix of signal values, described above, using the <em>computeMatrix</em> tool. To do so, you need to save the resulting boundaries as a .BED file (without a header), where the first column is the chromosome, second column is the boundary coordinate, and third column is (boundary coordinate + 1). The TSS option is used such that the 2nd column of the .BED file is chosen as the center (i.e. the boundary coordinate). Users will next need to download bigwig files for each cell-line specific annotation of interest, which can be done through ENCODE <a href="https://www.encodeproject.org/chip-seq-matrix/?type=Experiment&amp;replicates.library.biosample.donor.organism.scientific_name=Homo%20sapiens&amp;assay_title=TF%20ChIP-seq&amp;status=released">here</a>. The annotation (.bigwig) along with the boundaries (.BED) are provided next in the python code. We next specify by how much to flank around the boundary. Here, we opt for 5 kb (5000 bases) on either side. Additionally, we specify to divide the flanked region into 50 base segments (i.e. (5000+5000)/50=1000/50=200 segments). Lastly, we provide the name of the resulting matrix (computeMatrix_output.gz).</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1"></a><span class="co"># Create the matrix of signal using computeMatrix</span></span>
<span id="cb13-2"><a href="#cb13-2"></a>computeMatrix <span class="op">--</span>referencePoint TSS <span class="op">\</span> <span class="co">#uses the start coordinate (i.e. the boundary)</span></span>
<span id="cb13-3"><a href="#cb13-3"></a>              <span class="op">-</span>S annotation.bigWig <span class="op">\</span> <span class="co">#cell-line specific annotation of interest</span></span>
<span id="cb13-4"><a href="#cb13-4"></a>              <span class="op">-</span>R boundaries.bed <span class="op">\</span> <span class="co">#TAD boundaries (either predicted or called)</span></span>
<span id="cb13-5"><a href="#cb13-5"></a>              <span class="op">-</span>a <span class="dv">5000</span> <span class="op">-</span>b <span class="dv">5000</span> <span class="op">\</span> <span class="co">#size of flank on left side (-a) and right side (-b) in bases</span></span>
<span id="cb13-6"><a href="#cb13-6"></a>              <span class="op">--</span>binSize <span class="dv">50</span> <span class="op">\</span> <span class="co">#number of bases to divide the flanked region by</span></span>
<span id="cb13-7"><a href="#cb13-7"></a>              <span class="op">-</span>o computeMatrix_output.gz <span class="co">#name of the output</span></span></code></pre></div>
<p>Users can then use the computeMatrix_output.gz to create both the enriched heatmap and profile plot as shown below.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode python"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1"></a><span class="co"># Creating Enriched Heatmap </span></span>
<span id="cb14-2"><a href="#cb14-2"></a>plotHeatmap <span class="op">-</span>m computeMatrix_output.gz <span class="op">\</span> <span class="co">#name of the output of computeMatrix </span></span>
<span id="cb14-3"><a href="#cb14-3"></a>            <span class="op">-</span>out EnrichedHeatmap.png <span class="op">\</span> <span class="co">#name of heatmap image</span></span>
<span id="cb14-4"><a href="#cb14-4"></a>            <span class="op">--</span>whatToShow <span class="st">"heatmap and colorbar"</span> <span class="op">\</span> <span class="co">#what to include in the image</span></span>
<span id="cb14-5"><a href="#cb14-5"></a>            <span class="op">--</span>regionsLabel <span class="st">" "</span> <span class="op">\</span> <span class="co">#row labels</span></span>
<span id="cb14-6"><a href="#cb14-6"></a>            <span class="op">--</span>xAxisLabel <span class="st">" "</span> <span class="op">\</span> <span class="co">#x-axis label</span></span>
<span id="cb14-7"><a href="#cb14-7"></a>            <span class="op">--</span>refPointLabel <span class="st">"Boundary"</span> <span class="op">\</span> <span class="co">#name of the center point</span></span>
<span id="cb14-8"><a href="#cb14-8"></a>            <span class="op">--</span>samplesLabel <span class="st">"TAD-caller"</span> <span class="op">\</span> <span class="co">#column label</span></span>
<span id="cb14-9"><a href="#cb14-9"></a>            <span class="op">--</span>heatmapHeight <span class="dv">20</span> <span class="op">\</span> <span class="co">#height dimension in cm</span></span>
<span id="cb14-10"><a href="#cb14-10"></a>            <span class="op">--</span>heatmapWidth <span class="dv">5</span> <span class="op">\</span> <span class="co">#width dimension in cm</span></span>
<span id="cb14-11"><a href="#cb14-11"></a>            <span class="op">--</span>dpi <span class="dv">300</span> <span class="op">\</span> <span class="co">#dpi of saved image</span></span>
<span id="cb14-12"><a href="#cb14-12"></a>            </span>
<span id="cb14-13"><a href="#cb14-13"></a><span class="co"># Creating profile plot</span></span>
<span id="cb14-14"><a href="#cb14-14"></a>plotProfile <span class="op">-</span>m computeMatrix_output.gz <span class="op">\</span> <span class="co">#name of the output of computeMatrix </span></span>
<span id="cb14-15"><a href="#cb14-15"></a>            <span class="op">-</span>out ProfilePlot.png <span class="op">\</span> <span class="co">#name of profile plot image</span></span>
<span id="cb14-16"><a href="#cb14-16"></a>            <span class="op">--</span>colors blue <span class="op">\</span> <span class="co">#color of signal curve</span></span>
<span id="cb14-17"><a href="#cb14-17"></a>            <span class="op">--</span>regionsLabel <span class="st">" "</span> <span class="op">\</span> <span class="co">#row labels</span></span>
<span id="cb14-18"><a href="#cb14-18"></a>            <span class="op">--</span>samplesLabel <span class="st">"TAD-caller"</span> <span class="op">\</span> <span class="co">#column label</span></span>
<span id="cb14-19"><a href="#cb14-19"></a>            <span class="op">--</span>refPointLabel <span class="st">"Boundary"</span> <span class="op">\</span> <span class="co">#name of the center point</span></span>
<span id="cb14-20"><a href="#cb14-20"></a>            <span class="op">--</span>plotHeight <span class="dv">7</span> <span class="op">\</span> <span class="co">#height dimension in cm</span></span>
<span id="cb14-21"><a href="#cb14-21"></a>            <span class="op">--</span>plotWidth <span class="dv">7</span> <span class="op">\</span> <span class="co">#width dimension in cm</span></span>
<span id="cb14-22"><a href="#cb14-22"></a>            <span class="op">--</span>dpi <span class="dv">300</span> <span class="op">\</span> <span class="co">#dpi of saved image</span></span></code></pre></div>
</div>
<div id="comparing-precisetad-with-arrowhead" class="section level2">
<h2 class="hasAnchor">
<a href="#comparing-precisetad-with-arrowhead" class="anchor"></a>Comparing <em>preciseTAD</em> with Arrowhead</h2>
<p>We extend Section 6.4 by using <em>preciseTAD</em> to predict TAD boundaries on all of chromosome 14 for both GM12878 and K562 cell lines. This can be done by simply replacing the <code>chromCoords</code> parameter with <code>NULL</code>. We have provided the boundaries predicted by <em>preciseTAD</em> <a href="https://github.com/dozmorovlab/preciseTADworkshop/tree/main/inst/extdata/preciseTAD_5kb">here</a> and the Arrowhead called boundaries <a href="https://github.com/dozmorovlab/preciseTADworkshop/tree/main/inst/extdata/Arrowhead_5kb">here</a>. When trained using Arrowhead ground truth boundaries at 5 kb, <em>preciseTAD</em> predicted a total of 440 and 243 TAD boundaries on CHR14 for GM12878 and K562 respectively (Table 3). For comparison, Arrowhead identified a total of 555 and 240 TAD boundaries on CHR14 for GM12878 and K562 respectively (Table 3). We see that even though <em>preciseTAD</em> predicted less boundaries than Arrowhead, CTCF signal is much more colocalized around <em>preciseTAD</em>-predicted TAD boundaries compared to Arrowhead-called TAD boundaries (Figure 6A). This can be attributed to Arrowhead’s inflation of called TADs at 5 kb. Additionally, <em>preciseTAD</em> boundaries are statistically significantly closer to CTCF sites (Figure 6B). These results indicate that <em>preciseTAD</em>-predicted boundaries better reflected the known biology of boundary formation.</p>
<table class="table">
<thead><tr class="header">
<th></th>
<th>Arrowhead</th>
<th>preciseTAD</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>GM12878</td>
<td>555</td>
<td>440</td>
</tr>
<tr class="even">
<td>K562</td>
<td>240</td>
<td>243</td>
</tr>
</tbody>
</table>
<p>Table 3. TAD boundaries either called (Arrowhead) or predicted (<em>preciseTAD</em>) on CHR14.</p>
<center>
<div class="figure">
<img src="figures/pt_vs_arrowhead.png" alt=""><p class="caption"><strong>Figure 6. Levels of CTCF enrichment around <em>preciseTAD</em>-predicted boundaries are greater than Arrowhead-called boundaries.</strong></p>
</div>
</center>
<p>Previous studies suggest that TAD boundaries are conserved across cell lines <span class="citation">(Dixon et al. 2012; Krefting, Andrade-Navarro, and Ibn-Salem 2018; Nora et al. 2012; Sexton et al. 2012)</span>. To assess the level of cross-cell-line conservation, we can evaluate the overlap between cell line-specific boundaries detected by <em>preciseTAD</em> and Arrowhead. To do so, <em>preciseTAD</em>-predicted and Arrowhead-called TAD boundaries on CHR14 were first flanked on both sides by 5 kb. Only 21% of boundaries were conserved between cell lines for Arrowhead (J=0.16) (Figure 7A), while 37% of <em>preciseTAD</em>-predicted boundaries were conserved between GM12878 and K562 cell lines (J=0.32) (Figure 7B). The higher level of conservation for <em>preciseTAD</em>-predicted boundaries further supports the notion of their higher biological relevance.</p>
<center>
<div class="figure">
<img src="figures/conservation.png" style="width:75.0%" alt=""><p class="caption"><strong>Figure 7. <em>preciseTAD</em>-predicted TAD boundaries are more conserved across cell lines.</strong></p>
</div>
</center>
</div>
<div id="cross-cell-line-prediction" class="section level2">
<h2 class="hasAnchor">
<a href="#cross-cell-line-prediction" class="anchor"></a>Cross cell-line prediction</h2>
<p>The <em>preciseTAD</em> algorithm can be extended to predict boundaries on cell lines that do not currently have high resolution HiC data using publicly available cell-line specific annotations. We highlight this technique by evaluating the following scenario (for CHR14): training on GM12878 and predicting boundaries on GM12878 (GM on GM) vs. training on K562 and predicting on GM12878 (K on GM). Using the <em>preciseTAD</em> framework, 74% (J=0.68) of predicted boundaries overlapped in the cross-cell-line prediction scenario (Figure 8A). Furthermore, boundaries predicted on unseen annotation data exhibit a similar level of enrichment for CTCF as did those trained and predicted on the same cell line (Figure 7B). These results indicate that <em>preciseTAD</em> pre-trained models can be successfully used to predict domain boundaries for cell lines lacking Hi-C data but for which genome annotation data is available.</p>
<center>
<div class="figure">
<img src="figures/cross_cell_line_overlap.png" style="width:75.0%" alt=""><p class="caption"><strong>Figure 8. Cross cell-line predicted boundaries strongly overlap with same-cell-line predicted boundaries.</strong></p>
</div>
</center>
</div>
</div>
<div id="conclusion" class="section level1">
<h1 class="hasAnchor">
<a href="#conclusion" class="anchor"></a>Conclusion</h1>
<p>In this workshop we demonstrate the usefulness of <em>preciseTAD</em>, an R package that provides users with a data-driven approach toward TAD boundary prediction. <em>preciseTAD</em> leverages a random forest (RF) classification model built on low-resolution domain boundaries obtained from domain calling tools, and high-resolution genomic annotations as the feature space. <em>preciseTAD</em> predicts the probability of each base being a boundary, and identifies the precise location of boundary regions and the most likely boundary points among them. We benchmarked <em>preciseTAD</em> against Arrowhead <span class="citation">(Durand, Robinson, et al. 2016)</span>, an established TAD-caller. We showed that, unlike Arrowhead, <em>preciseTAD</em> does not inflate the number of predicted boundaries, providing only the most biologically meaningful boundaries that demarcate regions of high inter-chromosomal interactions. <em>preciseTAD</em> boundaries were more enriched for known architectural transcription factors, such as CTCF. Likewise, <em>preciseTAD</em> boundaries were more conserved between GM12878 and K562 cell lines, a known feature among the 3D architecture of the human genome. Lastly, we show that cell-line-specific functional genomic annotation data (ChIP-seq) can be used to precisely predict domain boundaries across cell lines. Pre-trained models are publicly available on Bioconductor in the ExperimentHub package <em>preciseTADhub</em>, <a href="https://github.com/dozmorovlab/preciseTADhub">here</a>. This creates new opportunities for users to predict domain boundaries on cell lines without relying on high-resolution Hi-C data for which there is none available.</p>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-breiman2001random">
<p>Breiman, Leo. 2001. “Random Forests.” <em>Machine Learning</em> 45 (1): 5–32.</p>
</div>
<div id="ref-dixon2015chromatin">
<p>Dixon, Jesse R, Inkyung Jung, Siddarth Selvaraj, Yin Shen, Jessica E Antosiewicz-Bourget, Ah Young Lee, Zhen Ye, et al. 2015. “Chromatin Architecture Reorganization During Stem Cell Differentiation.” <em>Nature</em> 518 (7539): 331.</p>
</div>
<div id="ref-dixon2012topological">
<p>Dixon, Jesse R, Siddarth Selvaraj, Feng Yue, Audrey Kim, Yan Li, Yin Shen, Ming Hu, Jun S Liu, and Bing Ren. 2012. “Topological Domains in Mammalian Genomes Identified by Analysis of Chromatin Interactions.” <em>Nature</em> 485 (7398): 376.</p>
</div>
<div id="ref-durand2016juicebox">
<p>Durand, Neva C, James T Robinson, Muhammad S Shamim, Ido Machol, Jill P Mesirov, Eric S Lander, and Erez Lieberman Aiden. 2016. “Juicebox Provides a Visualization System for Hi-c Contact Maps with Unlimited Zoom.” <em>Cell Systems</em> 3 (1): 99–101.</p>
</div>
<div id="ref-durand2016juicer">
<p>Durand, Neva C, Muhammad S Shamim, Ido Machol, Suhas SP Rao, Miriam H Huntley, Eric S Lander, and Erez Lieberman Aiden. 2016. “Juicer Provides a One-Click System for Analyzing Loop-Resolution Hi-c Experiments.” <em>Cell Systems</em> 3 (1): 95–98.</p>
</div>
<div id="ref-ester1996density">
<p>Ester, Martin, Hans-Peter Kriegel, Jörg Sander, Xiaowei Xu, and others. 1996. “A Density-Based Algorithm for Discovering Clusters in Large Spatial Databases with Noise.” In <em>Kdd</em>, 96:226–31. 34.</p>
</div>
<div id="ref-fudenberg2016formation">
<p>Fudenberg, Geoffrey, Maxim Imakaev, Carolyn Lu, Anton Goloborodko, Nezar Abdennur, and Leonid A Mirny. 2016. “Formation of Chromosomal Domains by Loop Extrusion.” <em>Cell Reports</em> 15 (9): 2038–49.</p>
</div>
<div id="ref-hahsler2019dbscan">
<p>Hahsler, Michael, Matthew Piekenbrock, and Derek Doran. 2019. “Dbscan: Fast Density-Based Clustering with R.” <em>Journal of Statistical Software</em> 25: 409–16.</p>
</div>
<div id="ref-hansen2018recent">
<p>Hansen, Anders S, Claudia Cattoglio, Xavier Darzacq, and Robert Tjian. 2018. “Recent Evidence That Tads and Chromatin Loops Are Dynamic Structures.” <em>Nucleus</em> 9 (1): 20–32.</p>
</div>
<div id="ref-hnisz2016activation">
<p>Hnisz, Denes, Abraham S Weintraub, Daniel S Day, Anne-Laure Valton, Rasmus O Bak, Charles H Li, Johanna Goldmann, et al. 2016. “Activation of Proto-Oncogenes by Disruption of Chromosome Neighborhoods.” <em>Science</em> 351 (6280): 1454–8.</p>
</div>
<div id="ref-krefting2018evolutionary">
<p>Krefting, Jan, Miguel A Andrade-Navarro, and Jonas Ibn-Salem. 2018. “Evolutionary Stability of Topologically Associating Domains Is Associated with Conserved Gene Regulation.” <em>BMC Biology</em> 16 (1): 1–12.</p>
</div>
<div id="ref-kuhn2012caret">
<p>Kuhn, Max. 2012. “The Caret Package.” <em>R Foundation for Statistical Computing, Vienna, Austria. URL Https://Cran. R-Project. Org/Package= Caret</em>.</p>
</div>
<div id="ref-liaw2002classification">
<p>Liaw, Andy, Matthew Wiener, and others. 2002. “Classification and Regression by randomForest.” <em>R News</em> 2 (3): 18–22.</p>
</div>
<div id="ref-lieberman2009comprehensive">
<p>Lieberman-Aiden, Erez, Nynke L Van Berkum, Louise Williams, Maxim Imakaev, Tobias Ragoczy, Agnes Telling, Ido Amit, et al. 2009. “Comprehensive Mapping of Long-Range Interactions Reveals Folding Principles of the Human Genome.” <em>Science</em> 326 (5950): 289–93.</p>
</div>
<div id="ref-lupianez2016breaking">
<p>Lupiáñez, Darı́o G, Malte Spielmann, and Stefan Mundlos. 2016. “Breaking Tads: How Alterations of Chromatin Domains Result in Disease.” <em>Trends in Genetics</em> 32 (4): 225–37.</p>
</div>
<div id="ref-Meaburn:2007aa">
<p>Meaburn, Karen J, Erik Cabuy, Gisele Bonne, Nicolas Levy, Glenn E Morris, Giuseppe Novelli, Ian R Kill, and Joanna M Bridger. 2007. “Primary Laminopathy Fibroblasts Display Altered Genome Organization and Apoptosis.” <em>Aging Cell</em> 6 (2): 139–53. <a href="https://doi.org/10.1111/j.1474-9726.2007.00270.x">https://doi.org/10.1111/j.1474-9726.2007.00270.x</a>.</p>
</div>
<div id="ref-nora2012spatial">
<p>Nora, Elphège P, Bryan R Lajoie, Edda G Schulz, Luca Giorgetti, Ikuhiro Okamoto, Nicolas Servant, Tristan Piolot, et al. 2012. “Spatial Partitioning of the Regulatory Landscape of the X-Inactivation Centre.” <em>Nature</em> 485 (7398): 381.</p>
</div>
<div id="ref-ramirez2016deeptools">
<p>Ramirez, Fidel, Devon P Ryan, Bjorn Gruning, Vivek Bhardwaj, Fabian Kilpert, Andreas S Richter, Steffen Heyne, Friederike Dundar, and Thomas Manke. 2016. “DeepTools2: A Next Generation Web Server for Deep-Sequencing Data Analysis.” <em>Nucleic Acids Research</em> 44 (W1): W160–W165.</p>
</div>
<div id="ref-rao20143d">
<p>Rao, Suhas SP, Miriam H Huntley, Neva C Durand, Elena K Stamenova, Ivan D Bochkov, James T Robinson, Adrian L Sanborn, et al. 2014. “A 3D Map of the Human Genome at Kilobase Resolution Reveals Principles of Chromatin Looping.” <em>Cell</em> 159 (7): 1665–80.</p>
</div>
<div id="ref-sanborn2015chromatin">
<p>Sanborn, Adrian L, Suhas SP Rao, Su-Chen Huang, Neva C Durand, Miriam H Huntley, Andrew I Jewett, Ivan D Bochkov, et al. 2015. “Chromatin Extrusion Explains Key Features of Loop and Domain Formation in Wild-Type and Engineered Genomes.” <em>Proceedings of the National Academy of Sciences</em> 112 (47): E6456–E6465.</p>
</div>
<div id="ref-Schmitt2016aa">
<p>Schmitt, Anthony D, Ming Hu, Inkyung Jung, Zheng Xu, Yunjiang Qiu, Catherine L Tan, Yun Li, et al. 2016. “A Compendium of Chromatin Contact Maps Reveals Spatially Active Regions in the Human Genome.” <em>Cell Rep</em> 17 (8): 2042–59. <a href="https://doi.org/10.1016/j.celrep.2016.10.061">https://doi.org/10.1016/j.celrep.2016.10.061</a>.</p>
</div>
<div id="ref-sexton2012three">
<p>Sexton, Tom, Eitan Yaffe, Ephraim Kenigsberg, Frédéric Bantignies, Benjamin Leblanc, Michael Hoichman, Hugues Parrinello, Amos Tanay, and Giacomo Cavalli. 2012. “Three-Dimensional Folding and Functional Organization Principles of the Drosophila Genome.” <em>Cell</em> 148 (3): 458–72.</p>
</div>
<div id="ref-Sun:2018aa">
<p>Sun, James H, Linda Zhou, Daniel J Emerson, Sai A Phyo, Katelyn R Titus, Wanfeng Gong, Thomas G Gilgenast, et al. 2018. “Disease-Associated Short Tandem Repeats Co-Localize with Chromatin Domain Boundaries.” <em>Cell</em> 175 (1): 224–238.e15. <a href="https://doi.org/10.1016/j.cell.2018.08.005">https://doi.org/10.1016/j.cell.2018.08.005</a>.</p>
</div>
<div id="ref-taberlay2016three">
<p>Taberlay, Phillippa C, Joanna Achinger-Kawecka, Aaron TL Lun, Fabian A Buske, Kenneth Sabir, Cathryn M Gould, Elena Zotenko, et al. 2016. “Three-Dimensional Disorganization of the Cancer Genome Occurs Coincident with Long-Range Genetic and Epigenetic Alterations.” <em>Genome Research</em> 26 (6): 719–31.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Spiro Stilianoudakis, Mikhail Dozmorov.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
